---
title: "Hogan Parentage"
output:
  html_notebook:
    theme: united
    toc: yes
editor_options: 
  chunk_output_type: inline
---

# Abstract
We are using genetic parentage analysis to estimate the relative reproductive success (RRS) of hatchery- and natural-origin pink salmon collected in Hogan Bay, a stream Prince William Sounds Alaska with annual escapements in the tens of thousands and observed stray rates ranging from 21-92% over the study period. We genotyped 8,000 hatchery and natural origin fish collected between 2013 and 2016 at 304 SNP amplicons. Standard quality control and quality assurance were performed, which included removing duplicate individuals, those who were successfully genotyped at <80% of markers, and those with abnormally high levels of heterozygosity. Population genetics analyses revealed that most of the variation was partitioned between lineages (odd versus even), but that there was little variation within lineages and between natural and hatchery origin individuals. *FRANz*, a Bayesian pedigree reconstructure program, produced parentage assignments for 2.5% of offspring genotyped for the odd lineage and 11% for the even lineage. Offspring were assigned to both natural- and hatchery-origin parents in both lineages. There were too few parentage assignments in the odd lineage to be able to robustly calculate RRS, but we were able to determine that natural origin females in the even lineage had significantly higher RRS than hatchery origin females and that there was not a significant difference between males. 


Below are analyses of parentage results calculated separately for Odd (2013-2015) and Even (2014-2016) lineages using *FRANz*. 

# Odd lineage (2013/2015)

## *FRANz* parameters
*FRANz* was run at 8:20AM on November 6, 2018. We used the following parameters:
FRANz.exe --Nmmax 23500 --Nfmax 23500 --femrepro 1:2 --malerepro 1:2 --typingerror 0.005 --updatefreqs --poutformat 2 --fullsibtest "V:\Analysis\5_Coastwide\Multispecies\Alaska Hatchery Research Program\PWS Pink\Franz\hogan_postQA_2013_2015_HOGAN.dat"

The parameters are defined as follows:
--Nmmax and --Nfmax are the maximum numbers of candidate mothers and fathers. To obtain our values, we used an estimated escapement of 47000 and divided by 2. This escapement number came from the larger of the two estimates (aerial survey AUC and stream walk AUC). This estimate is listed under 'pop est aerial' in the Excel file found here: "V:\Documents\5_Coastwide\Multispecies\AHRP\Field data\Sample Summary Extract and Genotype.xlsx"   
--femrepro and --malerepro specify the age range in which an individual can reproduce  
--typingerror refers to the overall genotyping error rate. Ours was ~0.005  
--updatefreqs specifies that *FRANz* should update allele frequencies using MCMC sampling  
--poutformat specifies that all potential parents should be listed, not just the most likely  
--fullsibtest tests for full siblings among offspring   

All output files can be found here: "v:\Analysis\5_Coastwide\Multispecies\Alaska Hatchery Research Program\PWS Pink\Franz\hogan_postQA_2013_2015_HOGAN_Run1"  

The summary file provides information about the power of our marker suite:   
Cumulative exclusion probability when 1 to 7 fullsibs are genotyped  
  First Parent              : 1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000   
  Second Parent             : 1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000   
  Parent Pair               : 1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000
  
According to the FRANz manual, marker sets are not considered powerful if these cumulative exclusion probabilities are less than 0.95, which indicates that the probability that a random pair of individuals in the population has a 5% chance of having a genotype pair compatible to an offspring genotype. Since all of our probabilities are 1, we can be **confident** in the power of our 304 amplicons to make parent assignments.  

## Import data
The first step is to read in .csv files for parentage assignments produced by *FRANz* as well as paired genotype and OceanAK data. 

```{r Setup Odd, message=FALSE, warning=FALSE}
library(tidyverse)
library(scales)
parentage_13_15 <- read_csv("v:/Analysis/5_Coastwide/Multispecies/Alaska Hatchery Research Program/PWS Pink/Franz/hogan_postQA_2013_2015_HOGAN_Run1/parentage.csv")
paired_13_15 <- read_csv("v:/Analysis/5_Coastwide/Multispecies/Alaska Hatchery Research Program/PWS Pink/Franz/hogan_postQA_OceanAK_paired_2013_2015_HOGAN.csv")
```

Plot a histogram of parentage posterior probabilities to show robustness of assignments
```{r FRANz_posterior_plot_odd}
parentage_13_15  %>% 
  filter(!is.na(`Parent 1`)) %>% 
  ggplot(aes(x = Posterior)) +
  geom_histogram(breaks = seq(0, 1, 0.01))
```
All parentage assigments have a posterior probability of 1, very robust.  

## Filter paired data
The file containing paired genotype and OceanAK data has a lot of information that we do not need at this time (e.g. genotypes for each marker). We therefore separate out only the columns that contain identifying information for each individual.

```{r Filter paired data Odd}
# What are the non-genotype columns
grep(pattern = "RAD", x = colnames(paired_13_15), value = TRUE, invert = TRUE)

# Filter for non-genotype columns
paired_13_15_filter <- paired_13_15 %>% 
  dplyr::select(franz_id, SILLY, `Fish ID`, `DNA Tray Code`, `DNA Tray Well Code`, `Sample Year`, `Sample Date`, SEX, `Length Mm`, `Otolith Mark Present`, `Otolith Mark ID`)

# Filter for posterior > 0.9 + make tidy relative to assigment (1 row per parentage assignment)
parentage_13_15_filter <- parentage_13_15 %>% 
  filter(Posterior > 0.9) %>% 
  dplyr::select(Offspring, `Parent 1`, `Parent 2`) %>% 
  gather(Parent, Parent_ID, - Offspring) %>% 
  filter(!is.na(Parent_ID))
```

## Sample sizes
Quick look at what our sample sizes are for potential parents (2013) and potential offspring (2015).

```{r odd_sample_sizes}
paired_13_15_filter %>% 
  mutate(origin = recode(`Otolith Mark Present`, !!!list("YES" = "hatchery", "NO" = "natural"))) %>% 
  count(`Sample Year`, SEX, origin) %>% 
  spread(origin, n, fill = 0)
```


## Join parentage and individual data
Here, we joined parentage data from *FRANz* with individual data so that we can match data to each offspring and parent. **Note** that any column ending in .x refers to offspring and .y refers to parent.  

```{r Join parentage and paired objects Odd}
# Creating a single, "tidy" object where each row is a parent-offspring relationship
# Offspring with 2 parents have 2 rows
# .x data is offspring, .y data is parents
(parents_paired_13_15 <- parentage_13_15_filter %>% 
   left_join(paired_13_15_filter, by=c("Offspring"="franz_id")) %>% 
   left_join(paired_13_15_filter, by=c("Parent_ID"="franz_id")))

parents_paired_13_15 %>% 
  count(Parent)

n_distinct(parents_paired_13_15$Parent_ID)

parents_paired_13_15 %>% 
  count(`Otolith Mark Present.y`, `Otolith Mark ID.y`) %>% 
  spread(`Otolith Mark Present.y`, n)
```
**Note** that *FRANz* assigned 48 offspring to 20 parents. 17 of these parents are natural origin and 3 are hatchery origin. The hatchery origin fish are from AFK11A, AFK11B, and WNH11PINKB.

Let's use a Chi-Square test to determine whether the proportions of natural and hatchery fish assigned as parents significantly differ from the proportions sampled. 

```{r Chi square test odd}
O.odd <- c("H", "N")
S.odd <- c(445, 322)
P.odd <- c(3, 17)

tbl.odd <- tibble(S.odd, P.odd) 
row.names(tbl.odd) <- c("H", "N")
View(tbl.odd)

chisq.test(tbl.odd)
```

There is a significant difference in the proportions of hatchery and natural fish sampled in the parental generation versus assigned as parents. 

## Family size
Here, we calculate and plot family size for both natural and hatchery parents.  Family size ranges from 1-16. All assignments were to 1 parent (*single parent-offspring duos*) - there were no 2 parent assignments (*parent pair-offspring trios*). 

Of the 1,920 offspring collected in 2015, 48 were assigned to parents, for an overall offspring assignment rate of **2.5%**. 

### Single parent-offspring pairs
This plot shows the distribution of family size by sex (number of offspring assigned to a female parent or a male parent) for both hatchery and natural parents.  

```{r Family size by sex Odd}
parents_paired_13_15 %>% 
  count(`Fish ID.y`, `Otolith Mark Present.y`, SEX.y) %>% 
  count(`Otolith Mark Present.y`, SEX.y, n) %>% 
  complete(`Otolith Mark Present.y`, SEX.y, n, fill = list(nn = 0)) %>% 
  ggplot(aes(x=n, y = nn, fill=`Otolith Mark Present.y`)) +
  geom_col(position = position_dodge2(preserve = "single")) +
  scale_y_continuous(breaks = pretty_breaks()) +
  scale_x_continuous(breaks = pretty_breaks()) +
  facet_grid(~ SEX.y) +
  labs(title="Distribution of Family Size by Females (F) and Males (M)",
       fill = "Parent: Otolith Mark Present") +
  xlab("Number of Offspring")+
  ylab("Number of Families") 

#parents_paired_13_15 %>% 
  # count(`Fish ID.y`, `Otolith Mark Present.y`, SEX.y) %>% 
  # count(`Otolith Mark Present.y`, SEX.y, n) %>% 
  # complete(`Otolith Mark Present.y`, SEX.y, n, fill = list(nn = 0)) %>% 
  # ggplot(aes()) +
  # stat_count(mapping = aes(x=n, y=..prop.., group=1, fill=`Otolith Mark Present.y`)) +
  # geom_col(position = position_dodge2(preserve = "single")) +
  # scale_y_continuous(breaks = pretty_breaks(), labels = percent_format()) +
  # scale_x_continuous(breaks = pretty_breaks()) +
  # facet_grid(~ SEX.y) +
  # labs(title="Distribution of Family Size by Females (F) and Males (M)",
  #      fill = "Parent: Otolith Mark Present") +
  # xlab("Number of Offspring")+
  # ylab("Number of Families")
```

When we plot the sexes separately, we can see that most of the assigned parents are males. Females produced 0-16 offspring, while males produced 0-4.  

```{r}
(rs_13_15 <- parents_paired_13_15 %>%
  count(`Fish ID.y`, `Otolith Mark Present.y`, SEX.y) %>%
  group_by(`Otolith Mark Present.y`, SEX.y) %>% 
  summarise(RS = mean(n))) %>% 
  spread(SEX.y, RS)
  

F_h_odd <- rs_13_15 %>% 
  filter(`Otolith Mark Present.y` == "YES" & SEX.y == "F") %>% 
  pull(RS)
M_h_odd <- rs_13_15 %>% 
  filter(`Otolith Mark Present.y` == "YES" & SEX.y == "M") %>% 
  pull(RS)
F_n_odd <- rs_13_15 %>% 
  filter(`Otolith Mark Present.y` == "NO" & SEX.y == "F") %>% 
  pull(RS)
M_n_odd <- rs_13_15 %>% 
  filter(`Otolith Mark Present.y` == "NO" & SEX.y == "M") %>% 
  pull(RS)
```


## Associations between parents and offspring
Here we briefly explore the heritability of:  

  * Body size
  * Body size x origin
  * Return timing
  * Return timing x origin

### Body size

Because body size is heritable, we compared length between parents and offspring.  

```{r Examine body length Odd}
parents_paired_13_15 %>% 
  filter(`Length Mm.y` != "NA") %>% 
  ggplot(aes(x=`Length Mm.y`, y=`Length Mm.x`, colour = `Otolith Mark Present.y`, shape = SEX.x))+
  geom_point(size = 2)+
  labs(title = "Offspring Length vs. Parent Length",
       colour = "Parent: Otolith Mark Present",
       shape = "Offspring: Sex") +
  xlab("Parent Length (mm)") +
  ylab("Offspring Length (mm)")
```

Offspring lengths show a great deal of variation. Male offspring tend to be larger than female offspring, as expected.  

```{r Fit a linear model between parent and offspring length Odd}
fit.lm.length <- lm(`Length Mm.x` ~ `Length Mm.y` + SEX.x, parents_paired_13_15)  
summary(fit.lm.length)
```

There is not a statistically significant relationship between parent and offspring length, but there is a significant effect of sex (p=0.03), which is what we observe in the scatterplot above, with males being larger on average than females. 

### Body size x origin

Let's examine the relationship between length and origin in the parent generation to see if there is a size difference between natural and hatchery origin fish. 

```{r Fit a linear model between parent length and origin Odd}
fit.lm.length.origin <- lm(`Length Mm`~ `Otolith Mark Present`+ SEX, paired_13_15_filter %>% filter(`Sample Year` == "2013"))
summary(fit.lm.length.origin)
```

Length significantly differs due to origin (p=0.010) and sex (p=0.019) in the parental generation (2013).

```{r Plot relationship between length and origin Odd}
paired_13_15_filter %>% 
  filter(`Sample Year` == "2013") %>% 
  ggplot(aes(x = SEX, y = `Length Mm`, fill = `Otolith Mark Present`)) +
  geom_violin() +
  theme(axis.text=element_text(size=6)) +
  ggtitle("Distribution of Length by Origin (2013)")
```

Hatchery fish have a tighter length distribution than natural origin fish, particularly in males. Below is a table of average length by otolith origin and sex for 2013 (the parental generation).

```{r Calculate mean lengths of hatchery and natural origin fish}
paired_13_15_filter %>% 
  filter(`Sample Year` == 2013) %>% 
  group_by(`Otolith Mark Present`, SEX) %>% 
  summarise(avg_length = round(mean(`Length Mm`, na.rm = TRUE))) %>% 
  spread(SEX, avg_length)
```
Hatchery males and females are larger than natural males and females. Both hatchery and natural females are larger than males. This is interesting and perhaps puzzling given that above we see that male offspring are larger than female offspring. Bear in mind that these were all fish collected in the parental generation and not just the parents. 

### Sample date

Run timing is also highly heritable, so we examined sample date in parents and their offspring. 

```{r make Julian date column}

parents_paired_13_15 <- 
parents_paired_13_15 %>% 
 mutate(Julian.x = yday(`Sample Date.x`), Julian.y = yday(`Sample Date.y`))
  


```


```{r Examine sample date Odd}
parents_paired_13_15 %>% 
  ggplot(aes(x=`Sample Date.y`, y=`Sample Date.x`, colour = `Otolith Mark Present.y`, shape = SEX.x))+
  geom_jitter(size = 2) +
    labs(title = "Offspring Date vs. Parent Date",
       colour = "Parent: Otolith Mark Present",
       shape = "Offspring: Sex") +
  xlab("Parent Sample Date") +
  ylab("Offspring Sample Date") 
```

Parents collected in early August had offspring who were also collected in August. Parents collected late August-early September had offspring collected across the duration of the season. Parents sampled later in September had offspring collected in late August-September (however, this is based on an n of 2). **Note** that hatchery-origin parents were not collected until late August-early September. Males also tend to be collected earlier than females.  

```{r Fit lm for parent versus offspring sample date}
lm.date.odd <- lm(parents_paired_13_15$Julian.x ~ parents_paired_13_15$Julian.y + parents_paired_13_15$SEX.x)
summary(lm.date.odd)
```

Offspring sample date has a significant association with parental sample date (p<0.001), but not offspring sex (p=0.248). 

### Sample date x origin

Let's compare sample dates across all of the individuals collected in 2013 to examine differences in run timing between hatchery and natural origin fish. 

```{r Compare sample dates by origin for 2013}
paired_13_15_filter %>% 
filter(`Sample Year`=="2013") %>% 
ggplot(aes(x=`Sample Date`, fill=`Otolith Mark Present`)) +
           geom_histogram(position=position_dodge()) +
  labs(title="Sample Date by Origin for 2013")

paired_13_15_filter %>% 
filter(`Sample Year`=="2015") %>% 
ggplot(aes(x=`Sample Date`)) +
           geom_histogram(position=position_dodge()) +
  labs(title="Sample Date for 2015")

paired_13_15_filter <- 
paired_13_15_filter %>% 
 mutate(Julian = yday(`Sample Date`))

paired_13_15_filter %>% 
  ggplot(aes(x=`Julian`, fill=`Otolith Mark Present`)) +
           geom_histogram(position=position_dodge()) +
  facet_wrap(~`Sample Year`, nrow=2) +
   labs(title="Sample Date")
```

From this histogram, we can see that in the parental generation (2013), natural origin fish tended to have earlier sampling dates than hatchery origin fish, which confirms what we know about differences in run timing. We can also see how sparse the field sampling was in 2013. Offspring in 2015 were sampled throughout the run.

### Length by sample date

```{r Plot length by sample date odd}
paired_13_15_filter %>% 
  filter(`Sample Year` == "2013") %>% 
  ggplot(aes(x = `Sample Date`, y = `Length Mm`, color = `Otolith Mark Present`)) +
  geom_jitter() +
  theme(axis.text=element_text(size=6)) +
  ggtitle("Distribution of Length by Sample Date (2013)")
```

This is a plot of parental sample date by length. Note that it is a jitter plot. 

```{r Fit LM for parental sample date and length}
lm.date.length.odd <- lm(parents_paired_13_15$`Length Mm.y`~parents_paired_13_15$`Sample Date.y`)
summary(lm.date.length.odd)

```

There is a significant relationship between parental sample date and length (p=0.033). 

```{r Calculate mean length by sample date odd}
parents_paired_13_15 %>% 
  filter(`Sample Year.y` == 2013) %>% 
  group_by(`Otolith Mark Present.y`, SEX.y, `Sample Date.y`) %>% 
  summarise(avg_length = round(mean(`Length Mm.y`, na.rm = TRUE))) %>% 
  spread(`Sample Date.y`, avg_length)
```

These are mean lengths by sex and origin for each sample date.  

```{r Length by sample date pooled odd}
parents_paired_13_15 %>% 
  filter(`Sample Year.y` == 2013) %>% 
  group_by(`Sample Date.y`) %>% 
  summarise(avg_length = round(mean(`Length Mm.y`, na.rm = TRUE))) %>% 
  spread(`Sample Date.y`, avg_length)
```

These are mean lengths by sample date with all individuals pooled. 

## Summary of odd lineage results
1. Exclusion probabilities were equal to 1, which means we can be confident in the ability of our marker set to correctly assign parents to offspring.  
2. 48 offspring were assigned to 20 parents (3 of which were hatchery-origin), for an assignment rate of **2.5%**.  
3. Family size varied from 0-16 for females and 0-4 for males.  
4. There was not a significant relationship between parent and offspring length.  
5. There is some evidence for a relationship between parent and offspring sample date (proxy for run timing) for parents collected at the beginning and end of the season.  
6. Hatchery fish appeared later in the run than natural fish. 


# Even lineage (2014/2016)

## *FRANz* parameters
*FRANz* was run at 10:30AM on November 6, 2018. We used the following parameters:
FRANz.exe --Nmmax 4500 --Nfmax 4500 --femrepro 1:2 --malerepro 1:2 --typingerror 0.005 --updatefreqs --poutformat 2 --fullsibtest "V:\Analysis\5_Coastwide\Multispecies\Alaska Hatchery Research Program\PWS Pink\Franz\hogan_postQA_2014_2016_HOGAN.dat"

The parameters are defined as follows:
--Nmmax and --Nfmax are the maximum numbers of candidate mothers and fathers. To obtain our values, we used an estimated escapement of 9000 and divided by 2. This escapement number came from the larger of the two estimates (aerial survey AUC and stream walk AUC). This estimate is listed under 'pop est aerial' in the Excel file found here: "V:\Documents\5_Coastwide\Multispecies\AHRP\Field data\Sample Summary Extract and Genotype.xlsx"   
--femrepro and --malerepro specify the age range in which an individual can reproduce  
--typingerror refers to the overall genotyping error rate. Ours was ~0.005  
--updatefreqs specifies that *FRANz* should update allele frequencies using MCMC sampling  
--poutformat specifies that all potential parents should be listed, not just the most likely  
--fullsibtest tests for full siblings among offspring   

All output files can be found here: "v:\Analysis\5_Coastwide\Multispecies\Alaska Hatchery Research Program\PWS Pink\Franz\hogan_postQA_2014_2016_HOGAN_Run1"  

The summary file provides information about the power of our marker suite:   
Cumulative exclusion probability when 1 to 7 fullsibs are genotyped  
  First Parent              : 1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000   
  Second Parent             : 1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000   
  Parent Pair               : 1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000  1.0000000
  
According to the FRANz manual, marker sets are not considered powerful if these cumulative exclusion probabilities are less than 0.95, which indicates that the probability that a random pair of individuals in the population has a 5% chance of having a genotype pair compatible to an offspring genotype. Since all of our probabilities are 1, we can be **confident** in the power of our 304 amplicons to make parent assignments.  

## Import data
The first step is to read in .csv files for parentage assignments produced by *FRANz* as well as paired genotype and OceanAK data.

```{r Setup Even, message=FALSE, warning=FALSE}
library(tidyverse)
parentage_14_16 <- read_csv("v:/Analysis/5_Coastwide/Multispecies/Alaska Hatchery Research Program/PWS Pink/Franz/hogan_postQA_2014_2016_HOGAN_Run 1/parentage.csv")
paired_14_16 <- read_csv("v:/Analysis/5_Coastwide/Multispecies/Alaska Hatchery Research Program/PWS Pink/Franz/hogan_postQA_OceanAK_paired_2014_2016_HOGAN.csv")
```

Plot a histogram of parentage posterior probabilities to show robustness of assignments
```{r FRANz_posterior_plot_even}
parentage_14_16  %>% 
  filter(!is.na(`Parent 1`)) %>% 
  ggplot(aes(x = Posterior)) +
  geom_histogram(breaks = seq(0, 1, 0.01))
```

Almost all parentage assigments have a posterior probability of 1, just looks like one offspring that is split between two pedigrees.

## Filter paired data
The file containing paired genotype and OceanAK data has a lot of information that we do not need at this time (e.g. genotypes for each marker). We therefore separate out only the columns that contain identifying information for each individual.  

Also filter for only parentage assignments with posterior probabilities > 0.9. This will ensure that we don't get offspring with two different single-parent assignments.

```{r Filter paired data Even}
# What are the non-genotype columns
grep(pattern = "RAD", x = colnames(paired_14_16), value = TRUE, invert = TRUE)

# Filter for non-genotype columns
paired_14_16_filter <- paired_14_16 %>% 
  dplyr::select(franz_id, SILLY, `Fish ID`, `DNA Tray Code`, `DNA Tray Well Code`, `Sample Year`, `Sample Date`, SEX, `Length Mm`, `Otolith Mark Present`, `Otolith Mark ID`)

# Filter for posterior > 0.9 + make tidy relative to assigment (1 row per parentage assignment)
parentage_14_16_filter <- parentage_14_16 %>% 
  filter(Posterior > 0.9) %>% 
  dplyr::select(Offspring, `Parent 1`, `Parent 2`) %>% 
  gather(Parent, Parent_ID, - Offspring) %>% 
  filter(!is.na(Parent_ID))
```

## Sample sizes
Quick look at what our sample sizes are for potential parents (2014) and potential offspring (2016).

```{r even_sample_sizes}
paired_14_16_filter %>% 
  mutate(origin = recode(`Otolith Mark Present`, !!!list("YES" = "hatchery", "NO" = "natural"))) %>% 
  count(`Sample Year`, SEX, origin) %>% 
  spread(origin, n, fill = 0)
```

## Join parentage and individual data
Here, we joined parentage data from *FRANz* with individual data so that we can match data to each offspring and parent. **Note** that any column ending in .x refers to offspring and .y refers to parent.  

```{r Join parentage and paired objects Even}
# Creating a single, "tidy" object where each row is a parent-offspring relationship
# Offspring with 2 parents have 2 rows
# .x data is offspring, .y data is parents
(parents_paired_14_16 <- parentage_14_16_filter %>% 
   left_join(paired_14_16_filter, by=c("Offspring"="franz_id")) %>% 
   left_join(paired_14_16_filter, by=c("Parent_ID"="franz_id")))

parents_paired_14_16 %>% 
  count(Parent)

n_distinct(parents_paired_14_16$Parent_ID)

parents_paired_14_16 %>% 
  summarise(count_h = n_distinct(Parent_ID[`Otolith Mark Present.y`=="YES"]), count_n = n_distinct(Parent_ID[`Otolith Mark Present.y`=="NO"]))

parents_paired_14_16 %>% 
  count(`Otolith Mark Present.y`, `Otolith Mark ID.y`) %>% 
  spread(`Otolith Mark Present.y`, n)
```
**Note** that *FRANz* assigned 456 offspring to 186 parents in single parent assignments (*single parent-offspring duos*). 209 of these assignments are to natural origin parents and 269 assignments were to hatchery origin parents. Hatchery origin parents are from AFK12B, CCH12, and WNH12PINKB, none are from Solomon Gulch (VFDA).  

Additionally, *FRANz* made 2-parent assignments (*parent pair-offspring trios*) for 22 offspring.


```{r Chi square test even}

S.even <- c(440, 114)
P.even <- c(214, 72)

tbl.even <- tibble(S.even, P.even) 
row.names(tbl.odd) <- c("H", "N")
View(tbl.even)

chisq.test(tbl.even)
```

There is not a significant difference between proportions of hatchery and natural origin fish sampled in the parental generation versus those assigned as parents. 

## Family size
Here, we calculate and plot family size for both natural and hatchery parents.  All potential cross types were present (natural-natural, hatchery-hatchery, hatchery-natural, natural-hatchery.  

Family size ranges from 1-13 for Parent 1 assignments and 1-3 for 2-parent assignments.  

Of the 4,027 offspring collected in 2016, 456 were assigned to at least 1 parent, for an an overall offspring assignment rate of **11%**.  

### Single parent-offspring pairs
This plot shows the distribution of family size by sex (number of offspring assigned to a female parent or a male parent) for both hatchery and natural parents.  

```{r Family size by sex Even}
parents_paired_14_16 %>% 
  count(`Fish ID.y`, `Otolith Mark Present.y`, SEX.y) %>% 
  count(`Otolith Mark Present.y`, SEX.y, n) %>% 
  complete(`Otolith Mark Present.y`, SEX.y, n, fill = list(nn = 0)) %>% 
  ggplot(aes(x=n, y = nn, fill=`Otolith Mark Present.y`)) +
  geom_col(position = position_dodge2(preserve = "single")) +
  scale_y_continuous(breaks = pretty_breaks()) +
  scale_x_continuous(breaks = pretty_breaks()) +
  facet_grid(~ SEX.y) +
  labs(title="Distribution of Family Size by Females (F) and Males (M)",
       fill = "Parent: Otolith Mark Present") +
  xlab("Number of Offspring")+
  ylab("Number of Families") 
```

For single parent-offspring assignments, females produced 1-10 offspring and males produced 1-13. For parent 2 assignments, females produced 1-3 offspring, while males produced 1-2. **This is likely to be a publication figure**.  

#### Plot reproductive success
Here we will calculate mean reproductive success by origin/sex and add it to the above plots.  

```{r calculate mean numbers of offspring}
(rs_14_16 <- parents_paired_14_16 %>%
  count(`Fish ID.y`, `Otolith Mark Present.y`, SEX.y) %>%
  group_by(`Otolith Mark Present.y`, SEX.y) %>% 
  summarise(RS = mean(n)))

F_h <- rs_14_16 %>% 
  filter(`Otolith Mark Present.y` == "YES" & SEX.y == "F") %>% 
  pull(RS)
M_h <- rs_14_16 %>% 
  filter(`Otolith Mark Present.y` == "YES" & SEX.y == "M") %>% 
  pull(RS)
F_n <- rs_14_16 %>% 
  filter(`Otolith Mark Present.y` == "NO" & SEX.y == "F") %>% 
  pull(RS)
M_n <- rs_14_16 %>% 
  filter(`Otolith Mark Present.y` == "NO" & SEX.y == "M") %>% 
  pull(RS)
```

Same plots as above, but show the mean reproductive success by origin.
```{r Plot mean offspring for females}
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

parents_paired_14_16 %>%
  filter(SEX.y == "F") %>% 
  count(`Fish ID.y`, `Otolith Mark Present.y`) %>% 
  count(`Otolith Mark Present.y`, n) %>% 
  complete(`Otolith Mark Present.y`, n, fill = list(nn = 0)) %>% 
  ggplot(aes(x=n, y = nn, fill=`Otolith Mark Present.y`)) +
  geom_col(position = position_dodge2(preserve = "single")) +
  scale_y_continuous(breaks = pretty_breaks()) +
  scale_x_continuous(breaks = pretty_breaks()) +
  xlab("Number of Offspring") +
  ylab("Number of Families") +
  labs(title="Number of Offspring for Females",
       fill = "Parent: Otolith Mark Present") +
  geom_vline(xintercept=F_n, size=2) +
  geom_vline(xintercept=F_n, color=gg_color_hue(2)[1], linetype = "dashed", size=2) +
  geom_vline(xintercept=F_h, size=2) +
  geom_vline(xintercept=F_h, color=gg_color_hue(2)[2], linetype = "dashed", size=2) +
  annotate("text", x = F_n + 0.5, y = 25, label = paste("F_n =", round(F_n, 2)), hjust = 0, size = 6, color = gg_color_hue(2)[1]) +
  annotate("text", x = F_h + 0.5, y = 30, label = paste("F_h =", round(F_h, 2)), hjust = 0, size = 6, color = gg_color_hue(2)[2]) 
```

These are the numbers of offspring for female hatchery and natural origin fish assigned as parent 1. Vertical lines represent mean family sizes (number of offspring) for natural and hatchery females. Mean offspring number for hatchery females = **`r round(F_h, 2)`**, while mean offspring number for natural females = **`r round(F_n, 2)`**.

```{r Plot mean offspring for males}
parents_paired_14_16 %>%
  filter(SEX.y == "M") %>% 
  count(`Fish ID.y`, `Otolith Mark Present.y`) %>% 
  count(`Otolith Mark Present.y`, n) %>% 
  complete(`Otolith Mark Present.y`, n, fill = list(nn = 0)) %>% 
  ggplot(aes(x=n, y = nn, fill=`Otolith Mark Present.y`)) +
  geom_col(position = position_dodge2(preserve = "single")) +
  scale_y_continuous(breaks = pretty_breaks()) +
  scale_x_continuous(breaks = pretty_breaks()) +
  xlab("Number of Offspring") +
  ylab("Number of Families") +
  labs(title="Number of Offspring for Males",
       fill = "Parent: Otolith Mark Present") +
  geom_vline(xintercept=M_n, size=2) +
  geom_vline(xintercept=M_n, color=gg_color_hue(2)[1], linetype = "dashed", size=2) +
  geom_vline(xintercept=M_h, size=2) +
  geom_vline(xintercept=M_h, color=gg_color_hue(2)[2], linetype = "dashed", size=2) +
  annotate("text", x = M_n + 0.5, y = 25, label = paste("M_n =", round(M_n, 2)), hjust = 0, size = 6, color = gg_color_hue(2)[1]) +
  annotate("text", x = M_h + 0.5, y = 30, label = paste("M_h =", round(M_h, 2)), hjust = 0, size = 6, color = gg_color_hue(2)[2]) 
```

This is a plot of number of offspring for natural and hatchery males assigned as parent 1. Vertical lines represent mean family sizes for natural and hatchery origin fish. Mean offspring number for male hatchery fish = **`r round(M_h, 2)`**, while mean offspring number for male natural fish = **`r round(M_n, 2)`**.

```{r Examine reproductive success by sample date}
(rs_14_16_date <- parents_paired_14_16 %>%
  count(`Fish ID.y`, `Otolith Mark Present.y`, SEX.y, `Sample Date.y`) %>%
  group_by(`Otolith Mark Present.y`, SEX.y, `Sample Date.y`) %>% 
  summarise(RS = mean(n)))

plot(rs_14_16_date$RS~rs_14_16_date$`Sample Date.y`, xlab="Parent")

rs_14_16_date %>% 
  ggplot(aes(x=`Sample Date.y`, y=RS, color=`Otolith Mark Present.y`, shape=SEX.y))+
  geom_point()+
  labs(title="RS By Parental Sample Date",  colour = "Parent: Otolith Mark Present",
       shape = "Offspring: Sex")+
  xlab("Parental Sample Date")+
  ylab("RS")


```

```{r LM to associate RS with parental sample date}
lm.rs.date <- lm(rs_14_16_date$RS~rs_14_16_date$`Sample Date.y`)
summary(lm.rs.date)
```

There is not a significant relationship between RS and sample date. 

#### Calculate relative reproductive success
We'll calculate relative reproductive success (RRS) using the following equation: mean hatchery family size/mean natural family size. We'll do this separately for males and females. We can get fancy and apply the unbiased RRS from Araki and Blouin 2005 later on.  

```{r Calculate relative reproductive success}
RRS_m <- M_h/M_n
RRS_f <- F_h/F_n

RRS_tibble <- tibble(RRS_m, RRS_f)
round(RRS_tibble, 2)
```

**RRS for males = `r round(RRS_m, 2)`. RRS for females = `r round(RRS_f, 2)`.**

Let's test for significant differences in RRS between male and female hatchery and natural origin fish. We'll perform two different statistical tests:  

  1. Nonparametric permutation test
  2. Parametric negative binomial GLM  
  
**Note** we are only doing this for the even lineage because the sample sizes are too low in the odd lineage. 

```{r RRS statistical testing}
# The format of the data to test was very simple, it was a data.frame called “mydata” with 2 columns
# 1.	nOff = number of offspring per family
# 2.	Origin = “H” or “W” for if the parent was hatchery or natural (wild)
# 

library(coin)
library(MASS)
library(beepr)
library(lattice)

mydata_M <- parents_paired_14_16 %>%
  filter(SEX.y == "M") %>% 
  count(`Fish ID.y`, `Otolith Mark Present.y`) %>% 
  mutate(`Otolith Mark Present.y` = as.factor(`Otolith Mark Present.y`))

mydata_F <- parents_paired_14_16 %>%
  filter(SEX.y == "F") %>% 
  count(`Fish ID.y`, `Otolith Mark Present.y`) %>% 
  mutate(`Otolith Mark Present.y` = as.factor(`Otolith Mark Present.y`))

perm_1tail_pvalue_M <- round(x = as.numeric(pvalue(oneway_test(n~`Otolith Mark Present.y`, data = mydata_M, distribution = approximate(B = 10000), alternative = "greater"))), digits = 4)

perm_1tail_pvalue_F <- round(x = as.numeric(pvalue(oneway_test(n~`Otolith Mark Present.y`, data = mydata_F, distribution = approximate(B = 10000), alternative = "greater"))), digits = 4)

#And to do a 1-tailed negative binomial GLM

fit_M <- glm.nb(n~`Otolith Mark Present.y`, data = mydata_M, init.theta = 1, link = log)
nbGLM_1tail_pvalue_M <- round(summary(fit_M)$coefficients[2, 4], 4)

fit_F <- glm.nb(n~`Otolith Mark Present.y`, data = mydata_F, init.theta = 1, link = log)
nbGLM_1tail_pvalue_F <- round(summary(fit_F)$coefficients[2, 4], 4)

(RRS <- tibble(perm_1tail_pvalue_F, perm_1tail_pvalue_M, nbGLM_1tail_pvalue_F, nbGLM_1tail_pvalue_M))
```

There **is** a significant difference in RS between hatchery- and natural-origin **females**, however, there **is not** any significant difference for **males**. 

#### Calculate RRS again including 0's 
We'll calculate RRS again including sampled fish from the parental generation to which offspring were not assigned.

```{r Re-calculate RRS}
RS_combined <- bind_rows(parents_paired_13_15, parents_paired_14_16) 
RS_combined_count <- RS_combined %>% 
  count(Parent_ID, `Otolith Mark Present.y`, SEX.y, `Sample Year.y`, `Length Mm.y`, `Sample Date.y`) %>% 
  mutate(lineage = dplyr::case_when(`Sample Year.y` %in% c("2013", "2015") ~ "odd",
                           `Sample Year.y` %in% c("2014", "2016") ~ "even"))

paired_14_16_filter_parents <- paired_14_16_filter %>% 
  filter(`Sample Year` == "2014")

paired_14_16_filter_parents <- left_join(paired_14_16_filter_parents, RS_combined_count, by = c("franz_id" = "Parent_ID")) %>% 
  mutate(n = replace_na(n, 0))

(rs_14_16_0 <- paired_14_16_filter_parents %>%
  group_by(`Otolith Mark Present`, SEX) %>% 
  summarise(RS = mean(n)))

paired_14_16_filter_parents %>% 
  ggplot(aes(x = n, fill = `Otolith Mark Present`)) +
  geom_bar(position = position_dodge2(preserve = "single")) +
  facet_grid( ~ SEX)


F_h_0 <- rs_14_16_0 %>% 
  filter(`Otolith Mark Present` == "YES" & SEX == "F") %>% 
  pull(RS)
M_h_0 <- rs_14_16_0 %>% 
  filter(`Otolith Mark Present` == "YES" & SEX == "M") %>% 
  pull(RS)
F_n_0 <- rs_14_16_0 %>% 
  filter(`Otolith Mark Present` == "NO" & SEX == "F") %>% 
  pull(RS)
M_n_0 <- rs_14_16_0 %>% 
  filter(`Otolith Mark Present` == "NO" & SEX == "M") %>% 
  pull(RS)
```

```{r Calculate relative reproductive success including 0s}
RRS_m_0 <- M_h_0/M_n_0
RRS_f_0 <- F_h_0/F_n_0

RRS_tibble_0 <- tibble(RRS_m_0, RRS_f_0)
round(RRS_tibble_0, 2)
```

**RRS for males = `r round(RRS_m_0, 2)`. RRS for females = `r round(RRS_f_0, 2)`.**

Let's test for significant differences in RRS between male and female hatchery and natural origin fish. We'll perform two different statistical tests:  

  1. Nonparametric permutation test
  2. Parametric negative binomial GLM  
  
**Note** we are only doing this for the even lineage because the sample sizes are too low in the odd lineage. 

```{r RRS statistical testing including 0s}
# The format of the data to test was very simple, it was a data.frame called “mydata” with 2 columns
# 1.	nOff = number of offspring per family
# 2.	Origin = “H” or “W” for if the parent was hatchery or natural (wild)
# 

library(coin)
library(MASS)
library(beepr)
library(lattice)

mydata_M_0 <- paired_14_16_filter_parents %>%
  filter(SEX == "M") %>% 
  #count(`Fish ID`, `Otolith Mark Present`) %>% 
  mutate(`Otolith Mark Present` = as.factor(`Otolith Mark Present`))

mydata_F_0 <- paired_14_16_filter_parents %>%
  filter(SEX == "F") %>% 
  #count(`Fish ID`, `Otolith Mark Present`) %>% 
  mutate(`Otolith Mark Present` = as.factor(`Otolith Mark Present`))

perm_1tail_pvalue_M_0 <- round(x = as.numeric(pvalue(oneway_test(n~`Otolith Mark Present`, data = mydata_M_0, distribution = approximate(B = 10000), alternative = "greater"))), digits = 4)

perm_1tail_pvalue_F_0 <- round(x = as.numeric(pvalue(oneway_test(n~`Otolith Mark Present`, data = mydata_F_0, distribution = approximate(B = 10000), alternative = "greater"))), digits = 4)

#And to do a 1-tailed negative binomial GLM

fit_M_0 <- glm.nb(n~`Otolith Mark Present`, data = mydata_M_0, init.theta = 1, link = log)
nbGLM_1tail_pvalue_M_0 <- round(summary(fit_M_0)$coefficients[2, 4], 4)

fit_F_0 <- glm.nb(n~`Otolith Mark Present`, data = mydata_F_0, init.theta = 1, link = log)
nbGLM_1tail_pvalue_F_0 <- round(summary(fit_F_0)$coefficients[2, 4], 4)

(RRS_0 <- tibble(perm_1tail_pvalue_F_0, perm_1tail_pvalue_M_0, nbGLM_1tail_pvalue_F_0, nbGLM_1tail_pvalue_M_0))
```

There is a significant difference in RRS for females, but not males. 

#### Tables
Of the offspring that we matched up to parents, how many went to hatchery-origin parents vs. natural-origin parents?
```{r Tables of parent information Even}
parents_paired_14_16 %>% 
  count(`Otolith Mark Present.y`)
```

269 offspring matched up to hatchery-origin parents, while 209 offspring assgined to natural-origin parents. 

### Parent pair-offspring trios

Because we have 22 individuals who were assigned to 2 parents, we are able to examine instances of crosses between hatchery fish, between natural fish, and between natural and hatchery fish. **Note** that we have divided natural-hatchery crosses into "HN", which indicates a hatchery origin mother and natural origin father and "NH", which indicates a natural origin mother and hatchery origin father. 

```{r Create cross as new grouping variable, message=FALSE}
# Which offspring have 2 parents?
offspring_trio_14_16 <- parents_paired_14_16 %>% 
  filter(Parent == "Parent 2") %>% 
  pull(Offspring)

parents_paired_14_16_cross <- parentage_14_16 %>% 
  filter(Posterior > 0.9) %>% 
  dplyr::select(Offspring, `Parent 1`, `Parent 2`) %>% 
  filter(Offspring %in% offspring_trio_14_16) %>% 
  left_join(paired_14_16_filter, by=c("Offspring"="franz_id")) %>% 
  left_join(paired_14_16_filter, by=c("Parent 1"="franz_id")) %>% 
  left_join(paired_14_16_filter, by=c("Parent 2"="franz_id")) %>% 
  dplyr::mutate(cross = dplyr::case_when(
                `Otolith Mark Present.y` == "NO" & `Otolith Mark Present` == "NO" ~ "NN",
                `Otolith Mark Present.y`== "YES" & `Otolith Mark Present` == "YES" ~ "HH",
                `Otolith Mark Present.y` == "YES" & SEX.y == "F" & `Otolith Mark Present` == "NO" & SEX == "M" ~ "HN",
                `Otolith Mark Present.y` == "NO" & SEX.y == "F" & `Otolith Mark Present` == "YES" & SEX == "M" ~ "NH",
                `Otolith Mark Present.y` == "YES" & SEX.y == "M" & `Otolith Mark Present` == "NO" & SEX == "F" ~ "NH",
                `Otolith Mark Present.y` == "NO" & SEX.y == "M" & `Otolith Mark Present` == "YES" & SEX == "F" ~ "HN" ))

parents_paired_14_16_cross %>% 
  count(cross, `Fish ID`, `Fish ID.y`) %>% 
  count(cross, n) %>% 
  complete(cross, n, fill = list(nn = 0)) %>% 
  ggplot(aes(x = n, y = nn, fill = cross))+
  geom_col(position=position_dodge2(preserve = "single")) +
  scale_y_continuous(breaks = pretty_breaks()) +
  scale_x_continuous(breaks = pretty_breaks()) +
  scale_x_continuous(breaks = c(1,2,3)) +
  labs(title="Distribution of Family Size by Cross") +
  xlab("Number of Offspring")+
  ylab("Number of Families") 
```

Offspring from all four types of crosses are represented in our dataset. HH crosses produced 1-2 offspring, HN and NH crosses produced 1-3 offspring, and NN crosses produced 1-2 offspring. Note that each fish only spawned with 1 other fish. 

## Associations between parents and offspring
Here we briefly explore the heritability of:  

  * Body size
  * Body size x origin
  * Return timing
  * Return timing x origin

### Body size

Because body size is heritable, we compared length between parents and offspring.  

```{r Examine body length Even}
parents_paired_14_16 %>% 
  filter(`Length Mm.y` != "NA") %>% 
  ggplot(aes(x=`Length Mm.y`, y=`Length Mm.x`, colour = `Otolith Mark Present.y`, shape = SEX.x))+
  geom_point(size = 2)+
  labs(title = "Offspring Length vs. Parent Length",
       colour = "Parent: Otolith Mark Present",
       shape = "Offspring: Sex") +
  xlab("Parent Length (mm)") +
  ylab("Offspring Length (mm)")
```

We fit linear models to examine relationships between offspring length and both parent length and offspring sex. 

```{r Fit a linear model between parent and offspring length Even}
fit.lm.length <- lm(`Length Mm.x` ~ `Length Mm.y` + SEX.x, parents_paired_14_16)  
summary(fit.lm.length)
```

Offspring length has a significant, albeit very small, linear relationship with parent 1 length (p=0.005), but there is not an effect of sex (p=0.58).  

### Body size x origin

We can also examine whether parent length differs between natural and hatchery origin fish in the parental generation. 

```{r Fit a linear model between parent length and origin Even}
fit.lm.length.origin <- lm(`Length Mm`~ `Otolith Mark Present`+ SEX, paired_14_16_filter %>% filter(`Sample Year` == "2014"))
summary(fit.lm.length.origin)
```

Length significantly differs due to origin (p=<0.001) and sex (p=<0.001) in the parental generation (2014). Let's plot it.  

```{r Plot relationship between length and origin Even}
paired_14_16_filter %>% 
  filter(`Sample Year` == "2014") %>% 
  ggplot(aes(x = SEX, y = `Length Mm`, fill = `Otolith Mark Present`)) +
  geom_violin() +
  theme(axis.text=element_text(size=6)) +
  ggtitle("Distribution of Length by Origin (2014)")
```

Male hatchery-origin fish have a tighter length distribution than natural-origin fish. Below is a table of average length by otolith origin and sex for BY 2014.

```{r Calculate mean lengths of hatchery and natural origin fish Even}
paired_14_16_filter %>% 
  filter(`Sample Year` == 2014) %>% 
  group_by(`Otolith Mark Present`, SEX) %>% 
  summarise(avg_length = round(mean(`Length Mm`, na.rm = TRUE))) %>% 
  spread(SEX, avg_length)
```
As we saw in the odd lineage, hatchery origin males and females are larger than natural origin males and females. Hatchery and natural origin females are larger than males. Bear in mind that these are all of the individuals collected in 2014, not just the parents.  

### Sample date

Because run timing is highly heritable, we want to examine relationships between parent and offspring sample date. We also examine sample dates of both parents in two parent families to see if they make sense (e.g. were they collected at about the same time). 

```{r Examine sample data Even}
parents_paired_14_16 %>% 
  ggplot(aes(x=`Sample Date.y`, y=`Sample Date.x`, colour = `Otolith Mark Present.y`, shape = SEX.x))+
  geom_jitter(size = 2) +
      labs(title = "Offspring Date vs. Parent Date",
       colour = "Parent: Otolith Mark Present",
       shape = "Offspring: Sex") +
  xlab("Parent Sample Date") +
  ylab("Offspring Sample Date") 

parents_paired_14_16_cross %>% 
  filter (SEX != "NA") %>% 
  ggplot(aes(x=`Sample Date.y`, y=`Sample Date`, colour = cross))+
  geom_point(size = 2) +
      labs(title = "Parent 2 Date vs. Parent 1 Date",
       colour = "Cross") +
  xlab("Parent 1 Sample Date") +
  ylab("Parent 2 Sample Date")
```

Natural origin fish appear to be sampled earlier in the season than hatchery origin fish. There is quite a bit of variation among offspring sample dates. As before, Otolith Mark Present refers to the parental generation and sex refers to offspring.  
Parent 1 and parent 2 sample dates closely align. All of the NN crosses occur early (before September 1). Crosses including hatchery fish occur throughout the run. However, it is interesting to note than natural-hatchery crosses and hatchery-natural crosses do not overlap. Hatchery-natural crosses (in which the female is hatchery and the male is natural) only occur at the end of the run. 

```{r Make Julian date column Even}


parents_paired_14_16 <- 
parents_paired_14_16 %>% 
 mutate(Julian.x = yday(`Sample Date.x`), Julian.y = yday(`Sample Date.y`))
  


```

```{r Fit lm for parent versus offspring sample date Even}
lm.date.even <- lm(parents_paired_14_16$Julian.x ~ parents_paired_14_16$Julian.y + parents_paired_14_16$SEX.x)
summary(lm.date.even)

```

Offspring sample date is significantly associated with parent sample date (p<0.001), but not sex (p=0.17). 

### Sample date x origin

```{r Plot proportion of hatchery fish by date Even}
paired_14_16_filter %>% 
filter(`Sample Year`=="2014") %>% 
ggplot(aes(x=`Sample Date`, fill=`Otolith Mark Present`)) +
  geom_histogram(position=position_dodge()) +
  labs(title="Sample Date by Origin for 2014")

paired_14_16_filter %>% 
filter(`Sample Year`=="2016") %>% 
ggplot(aes(`Sample Date`, fill=`Otolith Mark Present`)) +
  geom_histogram(position=position_dodge()) +
  labs(title="Sample Date for 2016")

paired_14_16_filter <- 
paired_14_16_filter %>% 
 mutate(Julian = yday(`Sample Date`))

paired_14_16_filter %>% 
  ggplot(aes(x=`Julian`, fill=`Otolith Mark Present`)) +
           geom_histogram(position=position_dodge()) +
  facet_wrap(~`Sample Year`, nrow=2) +
   labs(title="Sample Date")

```

These are sample dates for all of the fish in the parental generation (2014) and offspring generation (2016). In the parental generation, natural fish tended to be sampled earlier in the run and hatchery fish were more common later in the run. However, there is a high degree of overlap that facilitates the ability for natural and hatchery fish to interbreed. Most of the offspring were sampled in the second half of the season. 

### Length by sample date

```{r Plot length by sample date even}
paired_14_16_filter %>% 
  filter(`Sample Year` == "2014") %>% 
  ggplot(aes(x = `Sample Date`, y = `Length Mm`, color = `Otolith Mark Present`)) +
  geom_jitter() +
  theme(axis.text=element_text(size=6)) +
  ggtitle("Distribution of Length by Sample Date (2014)")
```

```{r Fit LM for parental sample date and length even}
lm.date.length.even <- lm(parents_paired_14_16$`Length Mm.y`~parents_paired_14_16$`Sample Date.y`)
summary(lm.date.length.even)

```

There is a significant relationship between sample date and length in the parental (2014) generation. 

```{r Calculate mean length by sample date even}
parents_paired_14_16 %>% 
  filter(`Sample Year.y` == 2014) %>% 
  group_by(`Otolith Mark Present.y`, SEX.y, `Sample Date.y`) %>% 
  summarise(avg_length = round(mean(`Length Mm.y`, na.rm = TRUE))) %>% 
  spread(`Sample Date.y`, avg_length)
```

This is a table of mean length for the parental generation (2014) by sampling date, broken down by sex and origin. 

```{r Length by sample date pooled even}
parents_paired_14_16 %>% 
  filter(`Sample Year.y` == 2014) %>% 
  group_by(`Sample Date.y`) %>% 
  summarise(avg_length = round(mean(`Length Mm.y`, na.rm = TRUE))) %>% 
  spread(`Sample Date.y`, avg_length)
```

These are mean lengths of all fish collected at each sample date. 

## Summary of even lineage results
1. Exclusion probabilities were equal to 1, which means we can be confident in the ability of our marker set to correctly assign parents to offspring.  
2. 456 offspring were assigned to 186 parents, for an assignment rate of **11%**. 209 assignments were to natural origin parents and 269 were to hatchery origin parents. *FRANz* was able to make both 1- and 2-parent assignments.      
3. Family size varied from 0-10 for females and 0-13 for males.  
4. Natural origin females have significantly higher RS than hatchery origin females, but there was not a significant difference for males.  
5. Hatchery and natural fish interbreed (both sexes).  
6. There was a significant relationship between parent and offspring length.  
7. Hatchery fish appeared later in the run than natural fish.

# ANOVA of RS across lineages

```{r ANOVA of RS across lineages}
 

aov.rs <- aov(RS_combined_count$n ~ RS_combined_count$lineage * RS_combined_count$`Otolith Mark Present.y` * RS_combined_count$SEX.y * RS_combined_count$`Length Mm.y` * RS_combined_count$`Sample Date.y`)
summary(aov.rs)

```

Orogin, lineage*sex, lineage*sex*length, lineage*length*date, origin*length*date, and sex*length*date all have a significant effect on RS. 

Since our degrees of freedom are very low, we'll run the ANOVA again without the interaction terms. 

```{r ANOVA without interactions}
aov.rs.2 <- aov(RS_combined_count$n ~ RS_combined_count$lineage + RS_combined_count$`Otolith Mark Present.y` + RS_combined_count$SEX.y + RS_combined_count$`Length Mm.y` + RS_combined_count$`Sample Date.y`)
summary(aov.rs.2)
```

None of the variables significantly influence RS on their own. Note that origin is closest to being statistically significant. 

# Overall Summary

1. *FRANz* was able to assign offspring to parents for both the odd and even lineages. The odd lineage offspring assignment rate was ~2.5%, while the even lineage offspring assignment rate was ~11%. Our marker set has a high degree of power, so we can be confident about the assignments that have been made.  
2. Family size varies from 1-16 individuals for the odd lineage and 1-13 individuals for the even lineage. For the odd lineage, females had greater variation in family size, while for the even lineage males had greater variation. However, the difference was slight (13 offspring as compared to 10).  
3. For the **odd-year** lineage, we saw that male and female natural origin fish had higher average numbers of offspring than hatchery fish, however, given the small sample size of families, we did not calculate *RRS* or do any statistical significance testing.  
4. For the **even-year** lineage, there is **significant difference** in *RRS* betwen female hatchery- and natural-origin parents (RRS  = `r round(RRS_f, 2)`, pval < 0.005 for both parametric and non-paramteric tests). However, there **is no significant difference** in *RRS* between male hatchery- and natural-origin parents (RRS  = `r round(RRS_m, 2)`, pval > 0.05 for both parametric and non-paramteric tests).  
5. Parent length did not have a significant effect on offspring length for the odd lineage, but did for the even lineage (albeit very small). The odd lineage did have a significant sex effect, however, with males being larger than females.  
6. For both lineages, hatchery origin parents had a tighter distribution of length than natural origin parents. In both lineages, hatchery origin fish tended to return later than natural origin fish, which is consistent with our prior knowledge. Also in both lineages, offspring sample date varied widely by parental sample date.  
7. For the even lineage, *FRANz* was able to assign 22 individuals to 2 parents. We observe four types of crosses: natural-natural, hatchery-hatchery, natural-hatchery (in which the female is natural origin), and hatchery-natural (in which the female is hatchery origin). Since this is a small sample size, it's difficult to draw conclusions about differences in reproductive success among offspring. However, it does appear that crosses involving hatchery fish occur throughout the season, while crosses involving natural fish occur early on.  
8. We ran *FRANz* again for each dataset using escapement estimates from stream walks to inform our *Nmmax* and *Nfmax* parameters (*i.e. see how sensitive parentage assignments are to our estimate of parent year escapement*). Assignments were identical for the **odd** year lineage, while the re-analysis of the **even** year lineage added three additional parent assignments (459 instead of 456).  These results indicate that our marker panel is very robust and that pedigree results are not highly sensitive to the *Nmmax* and *Nfmax* parameters. This is fantastic news given that our estimates of escapement are not very precise (**odd**: run1 - Nmmax = Nfmax = 23500 vs. run 2 - Nmmax = Nfmax = 4300; **even**: run1 - Nmmax = Nfmax = 4500 vs. run 2 - Nmmax = Nfmax = 3300).
9. In order to get an idea of our false-positive rate, we attempted to analyze 2013 as parents with 2016 as offspring, however, *FRANz* ran out of memory and failed. This is likely due to the number of bogus pedigrees it had to create, since no fish would be remotely related to each other.
  
# Potential Next Steps

1. Use CKMRSim to further evaluate the power of our marker set.  
2. Validate our findings by using another parentage software package (e.g. Cervus).  
3. Examine sibling assignments produced by *FRANz* as well as other software packages (e.g. fullsnplings).
4. Apply the unbiased RRS calculation from Araki and Blouin (2005). 
5. Run Dan's simulated data through *FRANz* to estimate potential false positive and negative rates. 