---
title: "Gilmour 2016/2018 Exploratory Analysis with 298 Markers"
output:
  html_notebook:
    theme: united
    toc: yes
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
rm(list=ls())
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(lubridate)
library(DT)
library(abind)
bbind <- function(...) { abind(..., along = 3) }

.username <- "kgruenthal"
.password <- ""
source("~/R/Functions.GCL.R")
```

# Introduction

This is a quick exploratory analysis of the Gilmour 2016/2018 genotype data. Here we will:

  * Read in *LocusControl*
  * Read in all project genotypes
  * Read in paired field/otoltih data from [OceanAK](http://www.oceanak.adfg.alaska.gov/)
  * Join paired data from OceanAK
    + Will need to create data key for join (DWP barcode + "_" + DWP position number)
  * Determine *FailureRate*
  * Perform QA on project gentoypes
    + Removing fish missing >= 20% of genotypes
    + Remove within collection duplicates
    + Explore the use of a **heterozygosity** filter to remove individuals with bogus genotype calls
    + Save final, post-QA genotypes with paired data
    * Export genotypes as a *Genepop* file to feed in to *genepop2franz.GCL.r*
  
# Read in genotypes

Create *LocusControl* and read in genotypes by *silly*
```{r loki}
sillyvec <- ProjectSillys <- c("PGILMOUR16", "PGILMOUR18")
# LocusControl <- dget("../Objects/Gilmour/LocusControl.txt")
loci298 <- dget("../../Objects/Stockdale_13_14_15_16/loci298.txt")

# loci <- LocusControl$locusnames
# nalleles <- LocusControl$nalleles
# ploidy <- LocusControl$ploidy
# alleles <- LocusControl$alleles

# LOKI2R.GCL(sillyvec = sillyvec, username = .username, password = .password)  # didn't work, hit memory limits within LOKI ("PGA_AGGREGATE_LIMIT")
read_project_genotypes.GCL(sillyvec = sillyvec, loci = loci298, username = .username, password = .password) # Update 11/26/19, LOKI2R worked after Eric contacted IT

loci <- LocusControl$locusnames
nalleles <- LocusControl$nalleles
ploidy <- LocusControl$ploidy
alleles <- LocusControl$alleles

rm(.username, .password)
```


Save objects
```{r save_locus_control_sillys}
if(!dir.exists("../../Objects/Gilmour_16_18/")) {dir.create("../../Objects/Gilmour_16_18/")}
save_objects(objects = c("LocusControl", "loci", "sillyvec", "ProjectSillys"), path = "../../Objects/Gilmour_16_18/")

if(!dir.exists("../../Genotypes/Gilmour_16_18/")) {dir.create("../../Genotypes/Gilmour_16_18/")}
save_sillys(sillyvec = sillyvec, path = "../../Genotypes/Gilmour_16_18/")
```
# OceanAK

## Read data
Read in paired field and otolith data from *OceanAK*, join that data to the attributes table of each silly object.
```{r read_oceanak, message=FALSE}
og_names <- suppressMessages(names(read_csv(file = "../../OceanAK/PedigreeData_AHRP - Salmon Biological Data 2_PWS_2013-2018_no_otoliths.csv", progress = FALSE)))
oceanak <- read_csv(file = "../../OceanAK/AHRP Salmon Biological Data 20200129_1621.csv")
names(oceanak) <- og_names

oceanak %>% 
  count(`Sample Year`, `Location Code`) %>% 
  spread(`Sample Year`, n, fill = 0)
```

Subset
```{r}
gilmour_oceanak <- oceanak %>% 
  dplyr::filter(`Location Code` == "Gilmour Creek" & `Sample Year` %in% c(2016, 2018)) %>% 
  tidyr::unite(silly_source, c("Silly Code", "Fish ID"), sep = "_", remove = FALSE)
```

```{r}
gilmour_oceanak %>% 
  count(`Sample Year`, `Otolith Mark Present`) %>% 
  spread(`Otolith Mark Present`, n, fill = 0)
```


Verify all genotypes in OceanAK
```{r}
table(PGILMOUR16.gcl$attributes$SillySource %in% gilmour_oceanak$silly_source)
table(PGILMOUR18.gcl$attributes$SillySource %in% gilmour_oceanak$silly_source)
```

What attribute names do we have since we used `read_project_genotypes.GCL` rather than `LOKI2R.GCL`?
```{r}
names(PGILMOUR16.gcl$attributes)
```


## Join attributes table
Join *gilmour_oceanak* data to the attributes table of each silly. Already looked to verify that no tissues were missing or duplicated, this makes sense because the lab can't genotype tissue that aren't there and they never genotype two samples in a well.
```{r join_oceanak_stock}
PGILMOUR16.gcl$attributes <- PGILMOUR16.gcl$attributes %>% 
  dplyr::left_join(gilmour_oceanak, by = c("SillySource" = "silly_source")) %>% 
  dplyr::select(FK_FISH_ID, SILLY_CODE, PLATE_ID, `DNA Tray Code`, `DNA Tray Well Code`, SillySource, `Sample Date`, Sex, `Length Mm`, `Location Code`, `Otolith Mark Present`, `Otolith Mark ID`, `Sample Year`) %>% 
  dplyr::rename(DNA_TRAY_CODE = "DNA Tray Code",
                DNA_TRAY_WELL_CODE = "DNA Tray Well Code")

PGILMOUR18.gcl$attributes <- PGILMOUR18.gcl$attributes %>% 
  dplyr::left_join(gilmour_oceanak, by = c("SillySource" = "silly_source")) %>% 
  dplyr::select(FK_FISH_ID, SILLY_CODE, PLATE_ID, `DNA Tray Code`, `DNA Tray Well Code`, SillySource, `Sample Date`, Sex, `Length Mm`, `Location Code`, `Otolith Mark Present`, `Otolith Mark ID`, `Sample Year`) %>% 
  dplyr::rename(DNA_TRAY_CODE = "DNA Tray Code",
                DNA_TRAY_WELL_CODE = "DNA Tray Well Code")

if(!dir.exists("../../Genotypes/Gilmour_16_18/OceanAK")) {dir.create("../../Genotypes/Gilmour_16_18/OceanAK")}
save_sillys(sillyvec = sillyvec, path = "../../Genotypes/Gilmour_16_18/OceanAK")
```


## Split by otolith-origin

First we need to verify that we only have hatchery and natural-origin fish for the **F0** year (2016), and only natural-origin fish for the **F1** year (2018). **However** at the moment we do not have full otolith reads for the 2016 or 2018 Gilmour samples, as they were part of the GOD event. We are going to just include all of them for now, as hatchery fish shouldn't assign to anyone.
```{r view_origin}
dplyr::bind_rows(PGILMOUR16.gcl$attributes, 
                 PGILMOUR18.gcl$attributes) %>% 
  count(SILLY_CODE, `Otolith Mark Present`) %>% 
  spread(`Otolith Mark Present`, n, fill = 0)
```

We want to remove any NA from 2016 and hatchery-origin fish, **but keep NA from 2016 and 2018**. We are going to leave all 2016 and 2018 fish in for now.
```{r remove_stock_hatchery}
# PGILMOUR16_ids_remove <- PGILMOUR16.gcl$attributes %>% 
#   filter(is.na(`Otolith Mark Present`)) %>% 
#   mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
#   pull(FK_FISH_ID)
# RemoveIDs.GCL(silly = "PGILMOUR16", IDs = PGILMOUR16_ids_remove)

# PGILMOUR18_ids_remove <- PGILMOUR18.gcl$attributes %>% 
#   filter(`Otolith Mark Present` == "YES" | is.na(`Otolith Mark Present`)) %>% 
#   mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
#   pull(FK_FISH_ID)
# RemoveIDs.GCL(silly = "PGILMOUR18", IDs = PGILMOUR18_ids_remove)
```

Verify
```{r verify_origin}
# dplyr::bind_rows(PGILMOUR16.gcl$attributes, 
#                  PGILMOUR18.gcl$attributes) %>% 
#   count(SILLY_CODE, `Otolith Mark Present`) %>% 
#   spread(`Otolith Mark Present`, n, fill = 0)
```

We will split 2016 otolith-origin for QA tabulating purposes. Also create a new *sillyvec*.
```{r split_stock15}
# PGILMOUR16_ids_natural <- list("PGILMOUR16" = PGILMOUR16.gcl$attributes %>% 
#   filter(`Otolith Mark Present` == "NO") %>% 
#   mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
#   pull(FK_FISH_ID))
# PoolCollections.GCL(collections = "PGILMOUR16", loci = loci, IDs = PGILMOUR16_ids_natural, newname = "PGILMOUR16n")
# 
# PGILMOUR16_ids_hatchery <- list("PGILMOUR16" = PGILMOUR16.gcl$attributes %>% 
#   filter(`Otolith Mark Present` == "YES") %>% 
#   mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
#   pull(FK_FISH_ID))
# PoolCollections.GCL(collections = "PGILMOUR16", loci = loci, IDs = PGILMOUR16_ids_hatchery, newname = "PGILMOUR16h")
# 
# ProjectSillys_Origin <- c("PGILMOUR16n", "PGILMOUR16h", "PGILMOUR18")
# if(!dir.exists("../../Genotypes/Gilmour_16_18/OceanAK_Origin")) {dir.create("../../Genotypes/Gilmour_16_18/OceanAK_Origin")}
# save_sillys(ProjectSillys_Origin, "../../Genotypes/Gilmour_16_18/OceanAK_Origin/")
```


# Failure rate

Calculate and view the failure rate by silly, locus, and plate.
```{r failure_rate}
project <- "Gilmour_16_18" #this is required of the function; can be any name 
failure_rate <- FailureRate.GCL(sillyvec = ProjectSillys)
failure_rate_noplots <- failure_rate[1:4]
save_objects("failure_rate_noplots", "../../Objects/Gilmour_16_18")
failure_rate
```
We can see some very specific plate effects.

# QA

Go through our standard Quality Assurance (QA) process, then explore heterozygosity. Produce a matrix with sample sizes.
```{r qa_setup}
ProjectSillys_SampleSizes <- matrix(data = NA, nrow = length(ProjectSillys), ncol = 5, dimnames = list(ProjectSillys, c("Genotyped", "Missing", "Duplicate", "Heterozygosity", "Final")))
ProjectSillys_SampleSizes[, "Genotyped"] <- sapply(paste0(ProjectSillys, ".gcl"), function(x) get(x)$n)
```

## Missing

Remove individuals missing >=20% of genotypes (i.e. the 80% rule).
```{r qa_missing}
MissLoci <- RemoveIndMissLoci.GCL(sillyvec = ProjectSillys, proportion = 0.8) 
MissLoci  # character vector of fish removed
save_objects("MissLoci", path = "../../Objects/Gilmour_16_18/")
ProjectSillys_SampleSizes[, "Missing"] <-  ProjectSillys_SampleSizes[, "Genotyped"] - sapply(paste0(ProjectSillys, ".gcl"), function(x) get(x)$n)
```

## Duplicate

Remove duplicate individuals within the same collection. Typically we specify *duplicates* as a pair of individuals that share >=95% of genotypes. Once a pair of *duplicates* is identified, we keep the individual with the most genotypes and remove the other.
```{r qa_duplicate}
DuplicateCheck95MinProportion <- CheckDupWithinSilly.GCL(sillyvec = ProjectSillys, loci = loci, quantile = NULL, minproportion = 0.95)
DuplicateCheckReportSummary <- sapply(ProjectSillys, function(x) DuplicateCheck95MinProportion[[x]]$report, simplify = FALSE)
DuplicateCheckReportSummary
save_objects("DuplicateCheckReportSummary", path = "../../Objects/Gilmour_16_18/")
```

```{r remove all duplicates}
#RemovedDups <- RemoveDups.GCL(DuplicateCheck95MinProportion)
#save_objects("RemovedDups", path = "../Objects/")

# All duplicate IDs in to one tibble
dups <- bind_rows(DuplicateCheckReportSummary[sapply(DuplicateCheckReportSummary, is.data.frame)], .id = "silly") %>% 
  select(silly, ID1, ID2) %>% 
  gather(trash, id, -silly) %>% 
  select(-trash) %>% 
  arrange(silly, id)

# Separate vector of IDs for each silly
dup_16_ids <- filter(dups, silly == "PGILMOUR16") %>% pull(id)
# dup_14h_ids <- filter(dups, silly == "PGILMOUR16h") %>% pull(id)  # none
dup_18_ids <- filter(dups, silly == "PGILMOUR18") %>% pull(id)

# Save attributes for posterity
# ID silly sources for duplicate individuals
dup_16 <- PGILMOUR16.gcl$attributes %>% 
  filter(FK_FISH_ID %in% dup_16_ids) %>% 
  mutate(FK_FISH_ID = as.numeric(FK_FISH_ID))
dup_18 <- PGILMOUR18.gcl$attributes %>%  
  filter(FK_FISH_ID %in% dup_18_ids)

# One tibble with all attributes for fish to remove
duplicates_to_remove <- bind_rows(dup_16, dup_18)
save_objects("duplicates_to_remove", path = "../../Objects/Gilmour_16_18/")

# Remove fish from .gcl objects
RemoveIDs.GCL(silly = "PGILMOUR16", IDs = dup_16_ids)
RemoveIDs.GCL(silly = "PGILMOUR18", IDs = dup_18_ids)

# Fill in sample size matrix
ProjectSillys_SampleSizes[, "Duplicate"] <-  ProjectSillys_SampleSizes[, "Genotyped"] - ProjectSillys_SampleSizes[, "Missing"] - sapply(paste0(ProjectSillys, ".gcl"), function(x) get(x)$n)
```

## Heterozygosity

We know from the QC work that the **GTscore** genotyping pipeline can sometimes result in individuals with an overabundance of heterozygous calls (perhaps from contamination?). We want to remove individuals with untrustworth genotypes. In order to explore the utility of some sort of *heterozygosity filter*, we should first plot a histogram of heterozygosity per individual to see what the problem looks like. From there we can set a threshold or range of acceptable heterozygosities.
```{r heterozygosity_calculations}
#genotype.counts <- apply(PHOGAN13.gcl$scores, MARGIN=1, table) #provides counts of each genotype/individual

#calculate heterozygosity by individual :-) 
PGILMOUR16_ind_het <- apply(PGILMOUR16.gcl$scores, 1, function(ind) {sum(ind[ ,"Dose1"] != ind[ ,"Dose2"])/298} ) 
#PGILMOUR16_ind_het <- apply(PGILMOUR16.gcl$scores, 1, function(ind) {sum(ind[ ,"Dose1"] != ind[ ,"Dose2"])/298} ) 
PGILMOUR18_ind_het <- apply(PGILMOUR18.gcl$scores, 1, function(ind) {sum(ind[ ,"Dose1"] != ind[ ,"Dose2"])/298} )
```

### Plot by silly
Plot with *tidyverse* for ease of use/visualization. Add +- 3SD cutoff from literature (Anderson et al. 2010).
```{r visualize heterozygosity by silly}
# Convert to tibbles
tibble_PGILMOUR16 <- tibble(silly=PGILMOUR16.gcl$attributes$SILLY_CODE, sillysource=PGILMOUR16.gcl$attributes$SillySource, het=PGILMOUR16_ind_het)
#tibble_PGILMOUR16h <- tibble(silly=PGILMOUR16h.gcl$attributes$SILLY_CODE, sillysource=PGILMOUR16h.gcl$attributes$SillySource, het=PGILMOUR16h_ind_het)
tibble_PGILMOUR18 <- tibble(silly=PGILMOUR18.gcl$attributes$SILLY_CODE, sillysource=PGILMOUR18.gcl$attributes$SillySource, het=PGILMOUR18_ind_het)

# Create a single tibble
het_tibble <- bind_rows(tibble_PGILMOUR16, tibble_PGILMOUR18)
 
#make het mean and lower and upper bounds (+/-3SD) tibbles
mean_het <- het_tibble %>% 
  dplyr::group_by(silly) %>% 
  dplyr::summarise(het=mean(het))
low_cutoff_het <- het_tibble %>% 
  dplyr::group_by(silly) %>% 
  summarise(het=mean(het)-3*sd(het))
high_cutoff_het <- het_tibble %>% 
  dplyr::group_by(silly) %>% 
  summarise(het=mean(het)+3*sd(het))
#plot individual heterozygosity by silly:
het_tibble %>% 
    ggplot2::ggplot(aes(x = het)) +
    ggplot2::geom_histogram(binwidth = 1/298) +
    ggplot2::geom_vline(aes(xintercept = het), mean_het) +
    ggplot2::geom_vline(aes(xintercept = het), low_cutoff_het) +
    ggplot2::geom_vline(aes(xintercept = het), high_cutoff_het) +
    ggplot2::facet_grid(rows=vars(silly), scales = "free_y") +
    ggplot2::xlim(0.15, 0.85) +
    ggplot2::xlab("Individual Heterozygosity") +
    ggplot2::ylab("Frequency") +
    ggplot2::ggtitle("Individual Heterozygosity By Silly")
```
After discussions with Chris, Tyler, and Dan, we decided that it would be more appropriate to examine heterozygosity by lineage (even versus odd) instead of year. Additionally, we all agreed that the +-3SD cutoff was not restrictive enough, the upper bound for the cutoff should be closer to the main distribution.

### Plot by lineage
```{r visualize heterozygosity by lineage}
#combine tibbles and add lineage as grouping variable
het_tibble <- het_tibble %>% 
  dplyr::mutate(lineage = dplyr::case_when(silly %in% c("PGILMOUR16", "PGILMOUR18") ~ "even",
                           TRUE ~ "even"))
save_objects("het_tibble", "../../Objects/Gilmour_16_18/")

#make het mean and lower and upper bounds (+/-3SD) tibbles
mean_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(het=mean(het))
low_cutoff_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  summarise(het=mean(het)-3*sd(het))
high_cutoff_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  summarise(het=mean(het)+3*sd(het))

#plot individual heterozygosity by lineage:
het_tibble %>% 
    ggplot2::ggplot(aes(x = het)) +
    ggplot2::geom_histogram(binwidth = 1/298) +
    ggplot2::geom_vline(aes(xintercept = het), mean_het) +
    ggplot2::geom_vline(aes(xintercept = het), low_cutoff_het) +
    ggplot2::geom_vline(aes(xintercept = het), high_cutoff_het) +
    ggplot2::facet_grid(rows=vars(lineage), scales = "free_y") +
    ggplot2::xlim(0.15, 0.85) +
    ggplot2::xlab("Individual Heterozygosity") +
    ggplot2::ylab("Frequency") +
    ggplot2::ggtitle("Individual Heterozygosity By Lineage")
```

How many fish would we lose if we went with these cutoffs?
```{r calculate how many individuals we would lose if we go with the 3SD cutoff}

# Upper cutoff only, lower doesn't apply
het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(upper = mean(het) + 3* sd(het)) %>% 
  dplyr::right_join(het_tibble) %>% 
  dplyr::filter(het >= upper) %>% 
  dplyr::count(lineage)
```
### Cutoffs to remove outliers
If we went with the 3SD cutoff, we would lose 223 samples from the odd lineage. Still, this +-3SD by lineage approach does not seem ideal given the large number of outliers that **pull** the SD. We can see this long right tail with a qqplot.

```{r make qq plots to visualize outliers}
#qq plot of heterozygosity by lineage:
het_tibble %>% 
    ggplot2::ggplot(aes(sample = het)) +
    ggplot2::stat_qq() +
    ggplot2::stat_qq_line() +
    ggplot2::facet_grid(rows = vars(lineage)) #+
    #ggplot2::coord_flip()
```
Same idea, but with a boxplot. **Note** all points are 1.5 IQR +- upper/lower boxplot. Perhaps we could use this 1.5 IQR approach for removing outliers?
```{r boxplots}
het_tibble %>% 
  ggplot2::ggplot(aes(x = lineage, y = het)) + 
  ggplot2::stat_boxplot(geom = "errorbar", width = 0.5) + 
  ggplot2::geom_boxplot(outlier.colour = "red", notch = FALSE, varwidth = FALSE, coef = 1.5) + 
  ggplot2::ggtitle("Boxplot of individual heterozygosity by lineage\noutliers are in red (1.5 IQR)")

# Calculate 1.5 IQR
het_IQR_cutoffs <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(low_1.5_IQR = quantile(het, probs = 0.25) - 1.5 * diff(quantile(het, probs = c(0.25, 0.75))),
            high_1.5_IQR = quantile(het, probs = 0.75) + 1.5 * diff(quantile(het, probs = c(0.25, 0.75))))

```
Chris suggested that we stick with the +-3SD approach, however, before calculating the SD, we mirror the left hand side of the distribution to the right (assuming the **mode** should be the **mean**).
```{r another way to visualize outliers (from Chris)}
# Calculate mode of each distribution
mode_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::count(het) %>% 
  dplyr::slice(which.max(n))


# Chris wants to assume the mode is the mean, mirror the lower half of the dist, calculate SD and use that for +-3SD
# Odd mirror lower bounds +-3sd
# het_tibble %>% 
#   dplyr::filter(lineage == "odd", het < 0.4078947) %>% 
#   dplyr::mutate(het = 0.4078947 - het + 0.4078947) %>% 
#   dplyr::bind_rows(het_tibble %>% 
#               filter(lineage == "odd", het <= 0.4078948)  # floating point error
#            ) %>% 
#   # ggplot(aes(x = het)) +
#   # geom_histogram(binwidth = 1/304)
#   dplyr::group_by(lineage) %>% 
#   dplyr::summarise(het_mean = mean(het), 
#             het_lower_3sd = mean(het) - 3 * sd(het),
#             het_upper_3sd = mean(het) + 3 * sd(het))


# Even mirror lower bounds +-3sd
het_tibble %>% 
  dplyr::filter(lineage == "even", het < 0.4161074) %>% 
  dplyr::mutate(het = 0.4177631 - het + 0.4161074) %>% 
  dplyr::bind_rows(het_tibble %>% 
              filter(lineage == "even", het <= 0.4161074)  # floating point error
  ) %>% 
  # ggplot(aes(x = het)) +
  # geom_histogram(binwidth = 1/304)
  dplyr::group_by(lineage) %>%
  dplyr::summarise(het_mean = mean(het),
            het_lower_3sd = mean(het) - 3 * sd(het),
            het_upper_3sd = mean(het) + 3 * sd(het))

```

Now lets look at all three of these approaches on the original histograms.

1. +-3SD by lineage
2. 1.5 IQR +- 75% and 25% quartiles
3. Chris' +-3SD by lineage only using the left hand side of the distribution.
```{r plot all of the different cutoffs on the same histogram.}
mean_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(het=mean(het))

low_cutoff_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(het=mean(het)-3*sd(het))

high_cutoff_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(het=mean(het)+3*sd(het))

low_box_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(het = quantile(het, probs = 0.25) - 1.5 * diff(quantile(het, probs = c(0.25, 0.75))))

high_box_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(het = quantile(het, probs = 0.75) + 1.5 * diff(quantile(het, probs = c(0.25, 0.75))))

odd_chris_het <- het_tibble %>% 
  dplyr::filter(lineage == "odd", het < 0.4078947) %>% 
  dplyr::mutate(het = 0.4078947 - het + 0.4078947) %>% 
  dplyr::bind_rows(het_tibble %>% 
              filter(lineage == "odd", het <= 0.4078948)  # floating point error
           ) %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(het_mean = mean(het), 
            het_lower_3sd = mean(het) - 3 * sd(het),
            het_upper_3sd = mean(het) + 3 * sd(het))

even_chris_het <-
  het_tibble %>% 
  dplyr::filter(lineage == "even", het < 0.4161074) %>% 
  dplyr::mutate(het = 0.4177631 - het + 0.4161074) %>% 
  dplyr::bind_rows(het_tibble %>% 
              filter(lineage == "even", het <= 0.4161074)  # floating point error
  ) %>% 
  dplyr::group_by(lineage) %>%
  dplyr::summarise(het_mean = mean(het),
            het_lower_3sd = mean(het) - 3 * sd(het),
            het_upper_3sd = mean(het) + 3 * sd(het))

chris_het_tibble <- dplyr::bind_rows(odd_chris_het, even_chris_het)

low_chris_het <- tibble(lineage=chris_het_tibble$lineage, het=chris_het_tibble$het_lower_3sd)

high_chris_het <- tibble(lineage=chris_het_tibble$lineage, het=chris_het_tibble$het_upper_3sd)

het_tibble %>% 
  ggplot2::ggplot(aes(x = het)) +
  ggplot2::geom_histogram(binwidth = 1/298) +
  ggplot2::geom_vline(aes(xintercept = het, color = "mean"), mean_het) +
  ggplot2::geom_vline(aes(xintercept = het, color = "SD"), low_cutoff_het) +
  ggplot2::geom_vline(aes(xintercept = het, color = "SD"), high_cutoff_het) +
  ggplot2::geom_vline(aes(xintercept = het, color = "IQR"), low_box_het) +
  ggplot2::geom_vline(aes(xintercept = het, color = "IQR"), high_box_het) +
  ggplot2::geom_vline(aes(xintercept = het, color = "Chris"), low_chris_het) +
  ggplot2::geom_vline(aes(xintercept = het, color = "Chris"), high_chris_het) +
  ggplot2::facet_grid(rows=vars(lineage), scales = "free_y") +
  ggplot2::scale_color_manual(name="statistics", values=c(mean="blue", SD="red", IQR="black", Chris="orange")) +
  ggplot2::xlim(0.15, 0.85) +
  ggplot2::xlab("Individual Heterozygosity") +
  ggplot2::ylab("Frequency") +
  ggplot2::ggtitle("Individual Heterozygosity By Lineage")

```
 
We can see that the upper bound is virtually identical for both Chris' method and IQR. We proposed that we go with the 1.5 IQR method given that it is well known and well documented. All agreed.

### Remove fish with IQR cutoff

Now that we've decided to use the 1.5IQR, how many fish are we going to remove from each silly?
```{r table_het_cutoff}
# Get list of fish to remove
het_tibble %>% 
  dplyr::left_join(het_IQR_cutoffs, by = "lineage") %>% 
  dplyr::filter(het < low_1.5_IQR | het > high_1.5_IQR) %>% 
  dplyr::count(silly)
```

Now remove those fish. Figure out which ones they are, convert *SillySource* to *FK_FISH_ID* (**note**: the values for *FK_FISH_ID* got reassigned after pooling), remove them.
```{r remove_het_cutoff}
removed_hets <- het_tibble %>% 
  dplyr::left_join(het_IQR_cutoffs, by = "lineage") %>% 
  dplyr::filter(het < low_1.5_IQR | het > high_1.5_IQR) %>% 
  dplyr::pull(sillysource)
save_objects("removed_hets", "../../Objects/Gilmour_16_18/")

# PGILMOUR16h_hets_remove <- PGILMOUR16h.gcl$attributes %>% 
#   dplyr::filter(SillySource %in% removed_hets) %>% 
#   dplyr::mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
#   dplyr::pull(FK_FISH_ID)
# RemoveIDs.GCL(silly = "PGILMOUR16h", IDs = PGILMOUR16h_hets_remove)

PGILMOUR16_hets_remove <- PGILMOUR16.gcl$attributes %>% 
  dplyr::filter(SillySource %in% removed_hets) %>% 
  dplyr::mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
  dplyr::pull(FK_FISH_ID)
RemoveIDs.GCL(silly = "PGILMOUR16", IDs = PGILMOUR16_hets_remove)

PGILMOUR18_hets_remove <- PGILMOUR18.gcl$attributes %>% 
  dplyr::filter(SillySource %in% removed_hets) %>% 
  dplyr::mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
  dplyr::pull(FK_FISH_ID)
RemoveIDs.GCL(silly = "PGILMOUR18", IDs = PGILMOUR18_hets_remove)

ProjectSillys_SampleSizes[, "Heterozygosity"] <-  ProjectSillys_SampleSizes[, "Genotyped"] - ProjectSillys_SampleSizes[, "Missing"] - ProjectSillys_SampleSizes[, "Duplicate"] - sapply(paste0(ProjectSillys, ".gcl"), function(x) get(x)$n)
```


## Final
How many fish did we end up with?
```{r qa_final}
ProjectSillys_SampleSizes[, "Final"] <- sapply(paste0(ProjectSillys, ".gcl"), function(x) get(x)$n)
ProjectSillys_SampleSizes
save_objects("ProjectSillys_SampleSizes", path = "../../Objects/Gilmour_16_18/")
```

Save post-QA genotypes.
```{r save_postQA_genotypes}
if(!dir.exists("../../Genotypes/Gilmour_16_18/OceanAK_PostQA")) {dir.create("../../Genotypes/Gilmour_16_18/OceanAK_PostQA")}
save_sillys(ProjectSillys, "../../Genotypes/Gilmour_16_18/OceanAK_PostQA/")
```


# Export genepop file

Now that we have final, post QA/QC genotypes, we want to save our work by exporting the genotypes as a **Genepeop** file. That **Genepop** file will then feed in to Chase's *genepop2franz.GCL.r* script to create a **FRANz** input file for parentage analysis!

Create a *genepop* using `gcl2Genepop.GCL`, individuals are named by *SillySource*.
```{r genepop}
if(!dir.exists("../../Genepop/Gilmour_16_18")) {dir.create("../../Genepop/Gilmour_16_18")}

gcl2Genepop.GCL(sillyvec = c("PGILMOUR16", "PGILMOUR18"), 
                loci = loci, 
                path = "../../Genepop/Gilmour_16_18/Gilmour_16_18_postQA.gen", 
                VialNums = TRUE, 
                usat = FALSE)
```

# Convert genepop to Franz

The `genepop2franz.GCL` function expects the *OceanAK* file to have the "old" names (lowercase with spaces), not the "new" names (UPPERCASE with _). Should convert to give the same rownames.

```{r Convert genepop to Franz for parentage}
detach(package:reshape, unload = TRUE)
source("../genepop2franz.GCL.r")
genepop2franz.GCL(
  Genepop = "../../Genepop/Gilmour_16_18/Gilmour_16_18_postQA.gen",
  OceanAK = "../../OceanAK/AHRP Salmon Biological Data 20200129_1621_names.csv",
  Year = c(2016, 2018),
  Stream = "GILMOUR",
  output_dir = "../"
)
```
End