---
title: "02 - Genotype QA Hearts & Otoliths"
subtitle: "Remove Untrustworthy Genotypes Prior to Matching"
author: "Kyle Shedd"
date: "2024-02-09 (hearts); 2024-05-22 (otoliths)"
output:
  html_notebook:
    theme: united
    toc: yes
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup

Load all necessary packages, as of 2023-08-11 Kyle switched to [GCLr](https://github.com/commfish/GCLr).
```{r setup, message=FALSE, results='hide'}
rm(list = ls(all.names = TRUE))

if(!require("pacman")) install.packages("pacman"); library(pacman)

pacman::p_load(
  tidyverse,
  lubridate,
  scales,
  janitor,
  GCLr
)

knitr::opts_chunk$set(fig.width = 10)

.username = readLines("~/R/usr_pw.txt", n = 1)  # LOKI username
.password = readLines("~/R/usr_pw.txt" , n = 2)[[2]]  # LOKI password
```

# Objective

The purpose of this notebook is to take the output from `01_import-genotypes-join-otlith-transfer-records.Rmd` and perform genotyping quality assurance (QA) on the otoliths in `PGOD21.gcl` and GOD-affected hearts in Erb 2015-2017, Gilmour 2015-2018, and Paddy 2016.

Hearts undergo standard AHRP PWS Pink Salmon QA:  

  1) missing genotypes (<80% loci)
  2) duplicate check (>95% loci, remove both)
  3) contamination (heterozygosity outliers, +/- 3.5 modified z-score)

Otoliths undergo a revised QA:  

  1) missing genotypes (<15% loci)
  2) duplicate check (>99% loci, remove 1)
  3) contamination (<20 non-heterozygous loci)

The next step after this will be to calculate the duplicate rate for each otolith-genotype compared to all heart-genotypes within a deep well plate (DWP).

# Background

During shipment of otolith samples in February 2020 to the MTA in Juneau for reading, a significant proportion of otoliths migrated between cells within DWPs due to poor containment of the acetate lids that were attached with rubber bands to the DWPs. This incident is known as "the great otolith debacle", aka the "GOD" incident. Since some otoliths moved from their original cells in the DWP, the paired integrity of the otolith-origin information and the rest of the paired data (genotype + field data) was lost for just under 11,000 individuals from the following collections:  

  * PERB15
  * PERB16
  * PERB17
  * PGILMOUR15
  * PGILMOUR16
  * PGILMOUR17
  * PGILMOUR18
  * PPADDY16

In an attempt to rectify the "GOD" incident, we extracted DNA from the otolith tissues, genotyped the otolith-derived DNA at 298 GT-seq loci, and are attempting to re-pair the otolith-heart samples from their genotypes.  

In addition to this GitHub repository, other materials can be found on the V: drive here:
[link](V:\Lab\Loki\Collection raw data sheets\Pink\Otolith Extravaganza\GOD Event 2020_2021)

# Methods

DNA was extracted from left-side or unknown otolith tissues of unknown origin using conventional Machery-Nagel DNA extraction kits. Otoliths were placed in T1 buffer for an overnight soak in clear 96 shallow-well plates (SWP). Preamp at 14 cycles. Final elution volume was 75uL. All otoliths were transferred among plates with a jig. All liquid handling was done by robot.

# Import Data

## Genotypes

### LocusControl

Using our standard 298 SNP pink salmon GT-seq loci.
```{r}
GCLr::load_objects(path = "../objects", pattern = "loci298")
GCLr::load_objects(path = "../objects", pattern = "LocusControl")
```

### Genotypes

Read in `PGOD21.gcl` and heart genotypes for all affected DWPs.
```{r}
GCLr::load_sillys(path = "../data/genotypes/raw_join_48DWP", rds = TRUE)
```

# Hearts

Standard data QA (GT-seq, no `conScore`):

  * Remove fish missing genotypes (**<80%** loci)
  * Remove duplicates (**>95%** loci concordance, remove **both** duplicates)
  * Remove contamination (heterozygosity outliers **+/- 3.5** modified z-score)

```{r}
(
  heart_sample_size_qa <- dplyr::tibble(silly = "heart_genotypes") %>%
    dplyr::mutate(genotyped = GCLr::silly_n(sillyvec = "heart_genotypes") %>% dplyr::pull(n))
)
```

## Missing Loci (<80%)

Remove fish with <80% loci genotyped.
```{r}
heart_miss_loci <-
  GCLr::remove_ind_miss_loci(sillyvec = "heart_genotypes",
                             proportion = 0.8)

# heart_miss_loci$IDs_Removed

GCLr::save_objects(objects = "heart_miss_loci",
                   path =  "../objects")

(
  heart_sample_size_qa <- heart_sample_size_qa %>%
    dplyr::mutate(missing = genotyped - GCLr::silly_n(sillyvec = "heart_genotypes") %>% dplyr::pull(n))
)
```

## Duplicate (>=95%)

Identify and remove duplicate genotypes (>=95% loci).
**NOTE** since we already subset down to heart samples involved in GOD, there is a slight chance that we may miss some lab duplicates since we aren't dealing with full plates of DNA. That said, I think this is fairly unlikely given that big duplicate issues would have been identified in lab QC.
```{r}
heart_duplicate_check_95 <-
  GCLr::dupcheck_within_silly(
    sillyvec = "heart_genotypes",
    minproportion = 0.95,
    minnonmissing = 0.6,
    ncores = 4
  )

GCLr::save_objects("heart_duplicate_check_95", path = "../objects")
```

Investigate duplicates a bit to see if they are from the same collections?
```{r}
heart_duplicate_check_95 %>% 
  dplyr::mutate(ID_diff = ID2 - ID1)
```

Yes, most seem to be nearest neighbor or close to each other.

Remove **both** duplicates! As opposed to GSI work, where we want to keep individuals but arenâ€™t typically worried about paired data, here we want to remove both individuals as the paired data integrity (including otolith reads!) is lost.
```{r}
heart_duplicates_removed <-
  GCLr::remove_dups(dupcheck = heart_duplicate_check_95, remove_both = TRUE)

GCLr::save_objects("heart_duplicates_removed", path = "../objects")

(
  heart_sample_size_qa <- heart_sample_size_qa %>%
    dplyr::mutate(
      duplicate = genotyped - missing - GCLr::silly_n(sillyvec = "heart_genotypes") %>% dplyr::pull(n)
    )
)
```

## Contamination (Heterozygosity)

Individual heterozygosity outliers is our best metric for removing contaminated samples in the absence of something more sophisticated like *conScore* from [GTscore](https://github.com/gjmckinney/GTscore?tab=readme-ov-file#sample-summaries). In previous analyses, including [Shedd et al. 2022](https://doi.org/10.1111/eva.13356), we've used the standard 1.5 IQR outlier detection method to remove excessively heterozygous individuals. However, the 1.5 IQR method assumes a normal distribution, but heart sample heterozygosities tend to be skewed right due to contamination. As of 2024-02-08, we decided to pivot to using [modified Z-scores](https://www.itl.nist.gov/div898/handbook/eda/section3/eda35h.htm) with cutoffs of +/- 3.5 as recommended by Iglewicz and Hoaglin [^1]. The modified Z-score uses the median and median absolute deviation (MAD) instead of the mean and standard deviation, and is thus more robust to outliers and asymmetrical distributions.

[^1]: Boris Iglewicz and David Hoaglin (1993), "Volume 16: How to Detect and Handle Outliers", The ASQC Basic References in Quality Control: Statistical Techniques, Edward F. Mykytka, Ph.D., Editor. 

Create a function to calculate individual heterozygosities.
```{r}
calc_ind_het <- function(sillyvec, loci = LocusControl$locusnames, ncores = 4) {
  
  start.time <- Sys.time() 
  
  if(!all(loci %in% LocusControl$locusnames)){
    
    stop(paste0("'", setdiff(loci, LocusControl$locusnames), "' from argument 'loci' not found in 'LocusControl' object!!!"))
    
  }
  
  if(!require("pacman")) install.packages("pacman"); library(pacman); pacman::p_load(tidyverse, doParallel, parallel, foreach)  # Install packages, if not in library and then load them
  
  if(ncores > parallel::detectCores()) {
    
    stop("'ncores' is greater than the number of cores available on machine\nUse 'detectCores()' to determine the number of cores on your machine")
    
  }
  
  all.gcl <- sapply(sillyvec, function(silly) {get(paste0(silly, ".gcl"), pos = 1)}, simplify = FALSE)
  
  cl <- parallel::makePSOCKcluster(ncores)
  
  doParallel::registerDoParallel(cl, cores = ncores)
  
  # Start parallel loop
  het <- foreach::foreach(silly = sillyvec, .packages = c("tidyverse")) %dopar% {
    
    my.gcl <- all.gcl[[silly]]
    
    dose1 <- my.gcl %>%
      dplyr::select(SillySource, dplyr::all_of(loci)) %>%
      tidyr::pivot_longer(
        cols = -SillySource,
        names_to = "locus",
        values_to = "dose1",
        values_drop_na = TRUE  # this drops any no-calls, which are NAs
      )
    
    dose2 <- my.gcl %>%
      dplyr::select(SillySource, dplyr::all_of(paste0(loci, ".1"))) %>%
      dplyr::rename_at(vars(paste0(loci, ".1")), ~ loci) %>%
      tidyr::pivot_longer(
        cols = -SillySource,
        names_to = "locus",
        values_to = "dose2",
        values_drop_na = TRUE  # this drops any no-calls, which are NAs
      )
    
    if(all.equal(dose1$SillySource, dose2$SillySource) & all.equal(dose1$locus, dose2$locus)) {
      
      my.gcl_tall <- dplyr::bind_cols(dose1, select(dose2, dose2))  # faster
      
    } else {
      
      my.gcl_tall <- dplyr::left_join(x = dose1, y = dose2, by = c("SillySource", "locus"))  # safer
      
    }
    
    my.gcl_tall %>%
      dplyr::group_by(SillySource) %>%
      dplyr::summarise(nloci = length(dose1),
                       ploci = nloci / length(loci),
                       het = sum(dose1 != dose2) / nloci)
    
  } %>%
    dplyr::bind_rows()
  
  parallel::stopCluster(cl)  # end parallel loop
  
  output <- het %>%
    tidyr::separate(
      col = SillySource,
      into = c("silly", "fish_id"),
      sep = "_",
      remove = FALSE
    ) %>% 
    dplyr::mutate(fish_id = as.numeric(fish_id)) %>% 
    dplyr::arrange(silly, fish_id)
  
  print(Sys.time() - start.time)
  
  return(output)
}
```

Calculate individual heterozygosities.
```{r}
(heart_ind_het <-
   calc_ind_het(
     sillyvec = "heart_genotypes",
     loci = loci298,
     ncores = 4
   ))
```

Need to split by lineage (even vs. odd) since allele frequencies are likely slightly different. Add in lineage and the *new* `FK_FISH_ID` from `heart_genotypes.gcl`.
```{r}
(
  heart_ind_het <- heart_ind_het %>%
    dplyr::mutate(
      year = 2000 + as.numeric(stringr::str_sub(
        string = silly,
        start = -2,
        end = -1
      )),
      lineage = dplyr::case_when(
        year %% 2 == 0 ~ "even",
        year %% 2 == 1 ~ "odd",
        TRUE ~ "mistakes_were_made"
      )
    ) %>%
    dplyr::left_join(
      y = heart_genotypes.gcl %>% dplyr::select(SillySource, FK_FISH_ID),
      by = "SillySource"
    )  # bring in FK_FISH_ID so we can use GCLr::remove_ids
)

GCLr::save_objects("heart_ind_het", path = "../objects")

heart_ind_het %>%
  dplyr::count(lineage, silly)
```

### Even

Plot distribution of modified z-scores of heterozygosity
```{r}
# Function to calculate median absolute deviation (MAD)
mad <- function(x) {
  median(abs(x - median(x)))
}

# Filter for even lineage & calculate the modified z-score for each individual
heart_ind_het_even <- heart_ind_het %>% 
  dplyr::filter(lineage == "even") %>%
  dplyr::mutate(modified_z_score = 0.6745 * (het - median(het)) / mad(het),
                outlier = dplyr::case_when(abs(modified_z_score) > 3.5 ~ TRUE,
                                           TRUE ~ FALSE))

heart_ind_het_even %>% 
  ggplot2::ggplot(ggplot2::aes(x = modified_z_score)) +
  ggplot2::geom_histogram(bins = length(loci298)) +
  ggplot2::facet_grid(rows = ggplot2::vars(lineage)) +
  ggplot2::geom_vline(xintercept = 0, colour = "blue") +
  ggplot2::geom_vline(xintercept = c(-3.5, 3.5), colour = "red") +
  ggplot2::xlab("Individual Heterozygosity Modified Z-Scores") +
  ggplot2::ylab("Frequency") +
  ggplot2::ggtitle("Individual Heterozygosity Modified Z-Scores - Even Lineage Collections") +
  ggplot2::theme_bw(base_size = 14) 
```

Plot as a distribution of heterozygosities showing both 1.5IQR cutoffs and outliers +/- 3.5 modified Z-score.
```{r}
het_1.5IQR_cutoffs_even <- heart_ind_het_even %>% 
  dplyr::summarise(
    het_median = median(het),
    het_1.5IQR_low = quantile(het, probs = 0.25) - 1.5 * diff(quantile(het, probs = c(0.25, 0.75))),
    het_1.5IQR_high = quantile(het, probs = 0.75) + 1.5 * diff(quantile(het, probs = c(0.25, 0.75)))
  )

heart_ind_het_even %>% 
  ggplot2::ggplot(ggplot2::aes(x = het, fill = outlier)) +
  ggplot2::geom_histogram(binwidth = 1 / length(loci298)) +
  ggplot2::facet_grid(rows = ggplot2::vars(lineage)) +
  ggplot2::geom_vline(data = het_1.5IQR_cutoffs_even, ggplot2::aes(xintercept = het_median, colour = "median")) +
  ggplot2::geom_vline(data = het_1.5IQR_cutoffs_even, ggplot2::aes(xintercept = het_1.5IQR_low, colour = "IQR")) +
  ggplot2::geom_vline(data = het_1.5IQR_cutoffs_even, ggplot2::aes(xintercept = het_1.5IQR_high, colour = "IQR")) +
  ggplot2::scale_color_manual(name = "statistics", values = c(median = "blue", IQR = "red")) +
  ggplot2::scale_fill_manual(name = "mod_z_score_outlier", values = c("TRUE" = "black", "FALSE" = "grey60")) +
  ggplot2::xlim(0, 1) +
  ggplot2::xlab("Individual Heterozygosity") +
  ggplot2::ylab("Frequency") +
  ggplot2::ggtitle("Individual Heterozygosity - Even Lineage Collections") +
  ggplot2::theme_bw(base_size = 14) 
```

How many fish from individual collections?
```{r}
heart_ind_het_even %>% 
  dplyr::count(silly, outlier) %>% 
  tidyr::pivot_wider(names_from = outlier, values_from = n)
```

Remove outliers based on +/- 3.5 modified Z-score.
```{r}
(
  heart_hets_removed_even <- GCLr::remove_ids(
    silly = "heart_genotypes",
    IDs = heart_ind_het_even %>% dplyr::filter(outlier == TRUE) %>% dplyr::pull(FK_FISH_ID)
  )
)

GCLr::save_objects("heart_hets_removed_even", path = "../objects")
```

### Odd

Plot distribution of modified z-scores of heterozygosity
```{r}
# Filter for odd lineage & calculate the modified z-score for each individual
heart_ind_het_odd <- heart_ind_het %>% 
  dplyr::filter(lineage == "odd") %>%
  dplyr::mutate(modified_z_score = 0.6745 * (het - median(het)) / mad(het),
                outlier = dplyr::case_when(abs(modified_z_score) > 3.5 ~ TRUE,
                                           TRUE ~ FALSE))

heart_ind_het_odd %>% 
  ggplot2::ggplot(ggplot2::aes(x = modified_z_score)) +
  ggplot2::geom_histogram(bins = length(loci298)) +
  ggplot2::facet_grid(rows = ggplot2::vars(lineage)) +
  ggplot2::geom_vline(xintercept = 0, colour = "blue") +
  ggplot2::geom_vline(xintercept = c(-3.5, 3.5), colour = "red") +
  ggplot2::xlab("Individual Heterozygosity Modified Z-Scores") +
  ggplot2::ylab("Frequency") +
  ggplot2::ggtitle("Individual Heterozygosity Modified Z-Scores - Odd Lineage Collections") +
  ggplot2::theme_bw(base_size = 14) 
```

Plot as a distribution of heterozygosities showing both 1.5IQR cutoffs and outliers +/- 3.5 modified Z-score.
```{r}
het_1.5IQR_cutoffs_odd <- heart_ind_het_odd %>% 
  dplyr::summarise(
    het_median = median(het),
    het_1.5IQR_low = quantile(het, probs = 0.25) - 1.5 * diff(quantile(het, probs = c(0.25, 0.75))),
    het_1.5IQR_high = quantile(het, probs = 0.75) + 1.5 * diff(quantile(het, probs = c(0.25, 0.75)))
  )

heart_ind_het_odd %>% 
  ggplot2::ggplot(ggplot2::aes(x = het, fill = outlier)) +
  ggplot2::geom_histogram(binwidth = 1 / length(loci298)) +
  ggplot2::facet_grid(rows = ggplot2::vars(lineage)) +
  ggplot2::geom_vline(data = het_1.5IQR_cutoffs_odd, ggplot2::aes(xintercept = het_median, colour = "median")) +
  ggplot2::geom_vline(data = het_1.5IQR_cutoffs_odd, ggplot2::aes(xintercept = het_1.5IQR_low, colour = "IQR")) +
  ggplot2::geom_vline(data = het_1.5IQR_cutoffs_odd, ggplot2::aes(xintercept = het_1.5IQR_high, colour = "IQR")) +
  ggplot2::scale_color_manual(name = "statistics", values = c(median = "blue", IQR = "red")) +
  ggplot2::scale_fill_manual(name = "mod_z_score_outlier", values = c("TRUE" = "black", "FALSE" = "grey60")) +
  ggplot2::xlim(0, 1) +
  ggplot2::xlab("Individual Heterozygosity") +
  ggplot2::ylab("Frequency") +
  ggplot2::ggtitle("Individual Heterozygosity - Odd Lineage Collections") +
  ggplot2::theme_bw(base_size = 14) 
```

How many fish from individual collections?
```{r}
heart_ind_het_odd %>% 
  dplyr::count(silly, outlier) %>% 
  tidyr::pivot_wider(names_from = outlier, values_from = n)
```

Remove outliers based on +/- 3.5 modified Z-score.
```{r}
(
  heart_hets_removed_odd <- GCLr::remove_ids(
    silly = "heart_genotypes",
    IDs = heart_ind_het_odd %>% dplyr::filter(outlier == TRUE) %>% dplyr::pull(FK_FISH_ID)
  )
)

GCLr::save_objects("heart_hets_removed_odd", path = "../objects")
```

```{r}
(
  heart_sample_size_qa <- heart_sample_size_qa %>%
    dplyr::mutate(
      heterozygosity = genotyped - missing - duplicate - GCLr::silly_n(sillyvec = "heart_genotypes") %>% dplyr::pull(n)
    )
)
```

## Final

```{r}
(
  heart_sample_size_qa <- heart_sample_size_qa %>%
    dplyr::mutate(
      final =  GCLr::silly_n(sillyvec = "heart_genotypes") %>% dplyr::pull(n)
    )
)

GCLr::save_objects("heart_sample_size_qa", path = "../objects")
readr::write_csv(x = heart_sample_size_qa, file = "../output/heart_sample_size_qa.csv")
```

Save the post-QA genotypes
```{r}
if(!dir.exists("../data/genotypes/postQA_join_48DWP")) {dir.create("../data/genotypes/postQA_join_48DWP")}

GCLr::save_sillys(
  sillyvec = "heart_genotypes",
  path = "../data/genotypes/postQA_join_48DWP",
  rds = TRUE
)
```

# Otoliths

Non-Standard data QA (GT-seq, no `conScore`):

  * Remove fish missing genotypes (**<15%** loci)
  * Remove duplicates (**>99%** loci concordance, remove **one** duplicate)
  * Remove contamination (**<20** non-heterozygous loci)

```{r}
(
  PGOD21_sample_size_qa <- dplyr::tibble(silly = "PGOD21") %>%
    dplyr::mutate(genotyped = GCLr::silly_n(sillyvec = "PGOD21") %>% dplyr::pull(n))
)
```

## Missing Loci (<15%)

Plot distribution of genotyping success (proportion loci genotyped) per otolith.
```{r}
PGOD21.gcl %>% 
  # dplyr::select(tidyselect::all_of(loci298)) %>% 
  dplyr::mutate(n_missing = rowSums(is.na(dplyr::pick(tidyselect::all_of(loci298)))),
                prop_loci = 1 - (n_missing / length(loci298))) %>% 
  ggplot2::ggplot(ggplot2::aes(x = prop_loci)) +
  ggplot2::geom_histogram(binwidth = 0.05) +
  ggplot2::theme_bw() +
  ggplot2::labs(x = "Proportion Loci Genotyped per Otolith",
                y = "Frequency",
                title = "PGOD21 Otolith Genotyping Success Rate")
```

Kristen had decided 15% previously, and that sounds better than 10%, but doesn't toss out too many otoliths.

Remove fish with <15% loci genotyped.
```{r}
PGOD21_miss_loci <-
  GCLr::remove_ind_miss_loci(sillyvec = "PGOD21",
                             proportion = 0.15)

# PGOD21_miss_loci$IDs_Removed

GCLr::save_objects(objects = "PGOD21_miss_loci",
                   path =  "../objects")

(
  PGOD21_sample_size_qa <- PGOD21_sample_size_qa %>%
    dplyr::mutate(missing = genotyped - GCLr::silly_n(sillyvec = "PGOD21") %>% dplyr::pull(n))
)
```

## Duplicate (>=99%)

Identify and remove duplicate genotypes (>=99% loci). This is higher than our standard 95% cutoff rate due to contamination and lower genotyping success for otoliths.
```{r}
PGOD21_duplicate_check_95 <-
  GCLr::dupcheck_within_silly(
    sillyvec = "PGOD21",
    minproportion = 0.99,
    minnonmissing = 0.6,
    ncores = 4
  )

GCLr::save_objects("PGOD21_duplicate_check_95", path = "../objects")
```

Investigate duplicates a bit to see if they are from the same collections?
```{r}
PGOD21_duplicate_check_95 %>% 
  dplyr::mutate(ID_diff = ID2 - ID1)
```

Yes, most seem to be nearest neighbor or close to each other, but some are very different!

Investigate genotypes of duplicate otoliths from disparate IDs
```{r}
PGOD21.gcl %>% 
  dplyr::filter(FK_FISH_ID %in% c(794, 4767, 2501, 5610, 2503, 5616))
```

Oh, interesting, these two otoliths are from the same DWP, but got sorted into different 96 SWP, so they have different FK_FISH_IDs. Otherwise, some of these duplicates are from vial/box otoliths that match up with 48_DWP_tray otoliths. I think that we should just remove the box/vial otolith and retain the 48_DWP_tray otoliths.
```{r}
PGOD21.gcl %>% 
  dplyr::filter(FK_FISH_ID %in% c(3902, 4084))
```

### Custom Duplicate Removal

Rather than toss BOTH duplicates, as in hearts, we will only toss ONE duplicate from each pair. We need to pair the otolith genotype (and it's read) to the heart genotyped to successfully re-pair the heart genotype and otolith read. Rather than keeping the otolith with the most loci genotyped (standard practice for GSI), we'd prefer to toss any box/vial otoliths, since they are not associated with a specific DWP. I briefly explored trying to join in otolith read data for the PGOD21 otoliths, but was unable to efficiently query the data and moved on.

Here's the logic for which otolith from each pair to toss  

  * If one of the duplicates is a box/vial otolith and the other is not, toss the box/vial  
  * If both DWP + position, toss whichever has the least amount of data  
  * If both vial/box, toss whichever has the least amount of data  

See what the different posibilities are.
```{r}
PGOD21.gcl %>% 
  dplyr::filter(FK_FISH_ID %in% c(PGOD21_duplicate_check_95$ID1, PGOD21_duplicate_check_95$ID2)) %>% 
  dplyr::count(`48_DWP_tray_position`, box_number)
```

Join in `48_DWP_tray_barcode` and `48_DWP_tray_position` from `PGOD21.gcl`, then assign *vial* or *DWP* based on `48_DWP_tray_position`, then add logic to preferentially select *vial* IDs to remove, otherwise standard whichever ID has the most missing loci.
```{r}
(
  PGOD21_duplicates_to_remove <- PGOD21_duplicate_check_95 %>%
    dplyr::left_join(
      y = PGOD21.gcl %>% dplyr::select(FK_FISH_ID, `48_DWP_tray_barcode`, `48_DWP_tray_position`),
      by = dplyr::join_by(ID1 == FK_FISH_ID)
    ) %>%
    dplyr::left_join(
      y = PGOD21.gcl %>% dplyr::select(FK_FISH_ID, `48_DWP_tray_barcode`, `48_DWP_tray_position`),
      by = dplyr::join_by(ID2 == FK_FISH_ID),
      suffix = c("1", "2")
    ) %>%
    dplyr::mutate(
      ID1_status = dplyr::case_when(
        is.na(`48_DWP_tray_position1`) |
          `48_DWP_tray_position1` == "vial" ~ "vial",
        TRUE ~ "DWP"
      ),
      ID2_status = dplyr::case_when(
        is.na(`48_DWP_tray_position2`) |
          `48_DWP_tray_position2` == "vial" ~ "vial",
        TRUE ~ "DWP"
      ),
      remove = dplyr::case_when(
        ID1_status == ID2_status & Missing1 > Missing2 ~ ID1,
        ID1_status == ID2_status &
          Missing2 > Missing1 ~ ID2,
        ID1_status == ID2_status &
          Missing1 == Missing2 ~ ID1,
        ID1_status != ID2_status &
          ID1_status == "vial" ~ ID1,
        ID1_status != ID2_status &
          ID2_status == "vial" ~ ID2
      )
    )
)

GCLr::save_objects("PGOD21_duplicates_to_remove", path = "../objects")
```

Remove **one** from each duplicate pair! In this case, we know that we have some duplicate otoliths since crews took both left and right otoliths.
```{r}
(
  PGOD21_duplicates_removed <- GCLr::remove_ids(silly = "PGOD21", IDs = PGOD21_duplicates_to_remove %>% dplyr::pull(remove))
)

GCLr::save_objects("PGOD21_duplicates_removed", path = "../objects")

(
  PGOD21_sample_size_qa <- PGOD21_sample_size_qa %>%
    dplyr::mutate(
      duplicate = genotyped - missing - GCLr::silly_n(sillyvec = "PGOD21") %>% dplyr::pull(n)
    )
)
```

## Contamination (Heterozygosity)

We may not know the lineage for all otoliths...does that matter? Let's look at the heart heterozygosity data broken out by lineage.
```{r}
heart_ind_het %>% 
  ggplot2::ggplot(ggplot2::aes(x = het)) +
  ggplot2::geom_histogram(binwidth = 0.01) +
  ggplot2::theme_bw() +
  ggplot2::facet_grid(rows = dplyr::vars(lineage)) +
  ggplot2::xlim(0, 1) +
  ggplot2::labs(x = "Individual Heterozygosity",
                y = "Frequency",
                title = "Heart Heterozygoisty by Lineage")
```

Overall, these distributions appear similar enough to me to justify analyzing all PGOD21 otoliths together. We do have some otoliths with unknown lineage (n = 142).
```{r}
PGOD21.gcl %>% dplyr::count(silly_code)
```

Calculate individual heterozygosities.
```{r}
(PGOD21_ind_het <-
   calc_ind_het(
     sillyvec = "PGOD21",
     loci = loci298,
     ncores = 4
   ))
```

Plot distribution of heterozygosity
```{r}
PGOD21_ind_het %>% 
  ggplot2::ggplot(ggplot2::aes(x = het)) +
  ggplot2::geom_histogram(binwidth = 0.01) +
  ggplot2::xlim(0, 1) +
  ggplot2::theme_bw() +
  ggplot2::labs(x = "Individual Heterozygosity",
                y = "Frequency",
                title = "Otolith Heterozygoisty (both Lineages)")
```

Whoa, this clearly demonstrates that the modified Z-score method will not work for the PGOD21 otoliths, there is just too much contamination!

Plot distribution of PGOD21 otolith vs. heart heterozygosity for context.
```{r}
heart_ind_het %>% 
  dplyr::bind_rows(PGOD21_ind_het %>% dplyr::select(-nloci, -ploci) %>% dplyr::mutate(lineage = "otolith")) %>% 
  ggplot2::ggplot(ggplot2::aes(x = het)) +
  ggplot2::geom_histogram(binwidth = 0.01) +
  ggplot2::xlim(0, 1) +
  ggplot2::theme_bw() +
  ggplot2::facet_grid(rows = dplyr::vars(lineage), scales = "free_y") +
  ggplot2::labs(x = "Individual Heterozygosity",
                y = "Frequency",
                title = "Heart vs. Otolith Heterozygoisty by Lineage")
```

Plot otolith het vs. genotyping success (proportion of loci genotyped). For hearts, our cutoff was 80% of loci, but for otoliths it was only 15%!!!
```{r}
PGOD21_ind_het %>% 
  ggplot2::ggplot(ggplot2::aes(x = ploci, y = het)) +
  ggplot2::geom_point() +
  ggplot2::theme_bw() +
  ggplot2::xlim(0, 1) +
  ggplot2::ylim(0, 1) +
  ggplot2::labs(x = "Genotype Rate (Proportion of Loci)",
                y = "Individual Heterozygosity",
                title = "Heterozygosity vs. Genotype Rate per Otolith")
```

Wow, quite the gamut of contamination here. We know that we had some contamination in T028 and that the heart-otolith duplicate rate matching method still worked, so that gives us some hope. That said, the actual PGOD21 otoliths appear to be much more heavily contaminated.

Rather than implement a simple heterozygosity cutoff, which doesn't account for the vast range of genotyping rates among individuals, I'm going to remove individuals with < 20 non-het genotypes. This is admittedly pretty arbitrary, but if we assume a MAF of ~ 0.35, the odds of any individual being homozygous for the major allele for 20 loci is > 100 times the odds of drawing 1 of the 10,791 PGOD21 otoliths that we started with at the beginning of QA. Here's the rough math.
```{r}
((1-0.35)^2)^20
1/10791
```

This is what removing individuals with < 20 non-heterozygous loci would look like.
```{r}
PGOD21_ind_het_to_remove <- PGOD21_ind_het %>% 
  dplyr::mutate(nhet = nloci * het,
                nhomo = nloci - nhet,
                remove = dplyr::case_when(nhomo < 20 ~ TRUE,
                                          TRUE ~ FALSE)) 

GCLr::save_objects("PGOD21_ind_het_to_remove", path = "../objects")

PGOD21_ind_het_to_remove %>% 
  ggplot2::ggplot(ggplot2::aes(x = ploci, y = het, colour = remove)) +
  ggplot2::geom_point() +
  ggplot2::theme_bw() +
  ggplot2::xlim(0, 1) +
  ggplot2::ylim(0, 1) +
  ggplot2::scale_color_manual(values = c("grey20", "red3")) +
  ggplot2::labs(x = "Genotype Rate (Proportion of Loci)",
                y = "Individual Heterozygosity",
                title = "Heterozygosity vs. Genotype Rate per Otolith - Remove < 20 non-Heterozygous Loci")
```

How many fish from individual collections?
```{r}
PGOD21_ind_het_to_remove %>%
  dplyr::left_join(
    y = PGOD21.gcl %>% dplyr::select(FK_FISH_ID, silly_code),
    by = dplyr::join_by(fish_id == FK_FISH_ID)
  ) %>%
  dplyr::count(silly_code, remove) %>%
  tidyr::pivot_wider(names_from = remove, values_from = n)
```

Remove outliers based on <20 non-heterozygous loci.
```{r}
(
  PGOD21_hets_removed <- GCLr::remove_ids(
    silly = "PGOD21",
    IDs = PGOD21_ind_het_to_remove %>% dplyr::filter(remove == TRUE) %>% dplyr::pull(fish_id)
  )
)

GCLr::save_objects("PGOD21_hets_removed", path = "../objects")
```

This doesn't seem like an unreasonable amount of otoliths to remove due to rampant contamination. This will help prevent us from chasing down matches that are based on spurious relationships.
```{r}
(
  PGOD21_sample_size_qa <- PGOD21_sample_size_qa %>%
    dplyr::mutate(
      heterozygosity = genotyped - missing - duplicate - GCLr::silly_n(sillyvec = "PGOD21") %>% dplyr::pull(n)
    )
)
```

## Final

```{r}
(
  PGOD21_sample_size_qa <- PGOD21_sample_size_qa %>%
    dplyr::mutate(
      final =  GCLr::silly_n(sillyvec = "PGOD21") %>% dplyr::pull(n)
    )
)

GCLr::save_objects("PGOD21_sample_size_qa", path = "../objects")
readr::write_csv(x = PGOD21_sample_size_qa, file = "../output/PGOD21_sample_size_qa.csv")
```

Save the post-QA genotypes
```{r}
if(!dir.exists("../data/genotypes/postQA_join_48DWP")) {dir.create("../data/genotypes/postQA_join_48DWP")}

GCLr::save_sillys(
  sillyvec = "PGOD21",
  path = "../data/genotypes/postQA_join_48DWP",
  rds = TRUE
)
```

# Summary

We started with 19,898 hearts from 8 collections. We removed 1,180 (5.9%) due to missing, 60 (0.6%) due to duplicates (both removed), and 592 (3.0%) due to contamination. Final set of hearts is 18,066 (90.8%).

We started with 10,791 otoliths in PGOD21. We removed 522 (4.8%) due to missing, 87 (0.8%) due to duplicates (only 1 removed), and 785 (7.3%) due to contamination. Final set of otoliths is 9,397 (87.1%).

End