---
title: "04 - Matching Hearts & Otoliths"
subtitle: "Assign Otolith Genotypes to Heart Genotypes by DWP"
author: "Kyle Shedd"
date: "Started: 2024-08-15, last opened: `r Sys.Date()`"
output:
  html_notebook:
    theme: united
    toc: yes
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup

Load all necessary packages, as of 2023-08-11 Kyle switched to [GCLr](https://github.com/commfish/GCLr).
```{r setup, message=FALSE, results='hide'}
rm(list = ls(all.names = TRUE))

if(!require("pacman")) install.packages("pacman"); library(pacman)

pacman::p_load(
  tidyverse,
  lubridate,
  scales,
  janitor,
  GCLr
)

knitr::opts_chunk$set(fig.width = 10)

.username = readLines("~/R/usr_pw.txt", n = 1)  # LOKI username
.password = readLines("~/R/usr_pw.txt" , n = 2)[[2]]  # LOKI password
```

# Objective

The purpose of this notebook is to make individual matching decisions for otolith- and heart-genotypes within a deep well plate (DWP) based on the duplicate genotype rate (% of loci sharing genotypes) for the top match and difference between the top and 2nd best match (`diff`).

Following that matching, we will need to consider whether any of the PGOD21 otoliths matched up to hearts/fish with "trustworthy", existing, otolith reads (i.e. we ended up with a right-side otolith in PGOD21, but the left-side otolith stayed put and was already read). Otherwise, we may have duplicate information for the same individual fish.

The next step after this will be to *compare matches to existing AHRP data to check for duplicates*. 

# Background

During shipment of otolith samples in February 2020 to the MTA in Juneau for reading, a significant proportion of otoliths migrated between cells within DWPs due to poor containment of the acetate lids that were attached with rubber bands to the DWPs. This incident is known as "the great otolith debacle", aka the "GOD" incident. Since some otoliths moved from their original cells in the DWP, the paired integrity of the otolith-origin information and the rest of the paired data (genotype + field data) was lost for just under 11,000 individuals from the following collections:  

  * PERB15
  * PERB16
  * PERB17
  * PGILMOUR15
  * PGILMOUR16
  * PGILMOUR17
  * PGILMOUR18
  * PPADDY16

In an attempt to rectify the "GOD" incident, we determined which otoliths were "suspect" and had likely migrated between wells, transferred those to 96 shallow well plates (SWP), extracted DNA from the otolith tissues, genotyped the otolith-derived DNA at 298 GT-seq loci, and are attempting to re-pair the otolith-heart samples from their genotypes. 

In addition to this GitHub repository, other materials can be found on the V: drive here:
[link](V:\Lab\Loki\Collection raw data sheets\Pink\Otolith Extravaganza\GOD Event 2020_2021)

# Methods

Here's a running list of methods and brief summary of results to help me remember how we got here since there are large gaps in between working on this project.

## Otolith Tranfers

We determined which otoliths were "suspect" and had likely migrated between wells based on comparing the DWP otolith inventory at the MTAL (destination, post-GOD) to the the inventory at the GCL (origin, pre-GOD). Full documentation of the *rules* are located here: `data/raw_working_oto_sheets/working_oto_sheet_combined.xlsm`, tab "decisions". Briefly, any well that was had an inventory discrepancy was "bad" and any adjacent well was "suspect" and physically labelled along the left edge of the well.

All otoliths from "suspect" wells were transferred to 96 SWPs (i.e., standard otolith trays) with 1 otolith per well. Full documentation of the *rules* are located in the Excel files here: `data/raw_otolith_transfer_96`, see tab "oto_transfer_business_rules". After completing the transfers at the MTAL, SWPs were sent to the GCL for DNA extraction and genotyping.

## Laboratory

DNA was extracted from `PGOD21` left-side or unknown otolith tissues of unknown origin using conventional Machery-Nagel DNA extraction kits. Otoliths were placed in T1 buffer for an overnight soak in clear 96 SWPs. Preamp at 14 cycles. Final elution volume was 75uL. All otoliths were transferred among plates with a jig. All liquid handling was done by robot. Samples were genotyped using GT-seq in lab projects P062, P065, P066, P067, P068, P069, P070, P071, P072, P073. QC samples were included in P067, P068, and P074.

## Analysis Workflow

### Data Prep

All data prep was done in the R Notebook `01_import-genotypes-join-otlith-transfer-records`. The purpose of that R Notebook was to come up with a definitive link between the `PGOD21.gcl` otoliths and their 96 SWP locations and the *original* 48 DWP or otolith box location. This was accomplished by linking these 3 things:  

  1) `PGOD21.gcl` otolith genotypes from 96 SWP,
  2) Otolith transfer records documenting the transfer of otoliths from their original 48 DWP (or box of 48 DWP),
  3) AHRP Salmon Biological Fact data containing field data (silly code, stream, year, 48 DWP, etc.)
  
Bottom line, we need to know to the finest scale possible (48 DWP or box) where otoliths in `PGOD21.gcl` came from so we can limit our search space for matching.

How many otoliths with genotypes in `PGOD21.gcl`? By collection?

Final stats for `PGOD21.gcl` from `01_import-genotypes-join-otlith-transfer-records`:

 silly_code        n
     PERB15      478
     PERB16    1,527
     PERB17      525
 PGILMOUR15    1,587
 PGILMOUR16    1,555
 PGILMOUR17    2,600
 PGILMOUR18      856
   PPADDY16    1,503
       Box       159
Unaccounted        1
      Total   10,791

match_type         n
          Box    159
 DWP+Position 10,201
     DWP_Only    430
       Nothin      1
        Total 10,791

**439 unique 48 DWPs**

### Quality Assurance (QA)

All QA was done in the R Notebook `02_genotype-QA-hearts-otoliths`. The purpose of that R Notebook was to perform genotyping quality assurance (QA) on the otoliths in PGOD21.gcl and GOD-affected hearts in Erb 2015-2017, Gilmour 2015-2018, and Paddy 2016.

Hearts undergo standard AHRP PWS Pink Salmon QA:  

  1) missing genotypes (<80% loci)
  2) duplicate check (>95% loci, remove both)
  3) contamination (heterozygosity outliers, +/- 3.5 modified z-score, separate for even/odd lineage)

Otoliths undergo a revised QA:  

  1) missing genotypes (<15% loci, arbitrary cutoff)
  2) duplicate check (>99% loci[contamination + low genotyping success], remove 1 [keep sample w/ most location info and genotype data])
  3) contamination (<20 non-heterozygous loci, all together not separate by lineage)

We started with 19,898 hearts from 8 collections. We removed 1,180 (5.9%) due to missing, 60 (0.6%) due to duplicates (both removed), and 592 (3.0%) due to contamination. Final set of hearts is 18,066 (90.8%).

We started with 10,791 otoliths in PGOD21. We removed 522 (4.8%) due to missing, 87 (0.8%) due to duplicates (only 1 removed), and 785 (7.3%) due to contamination. Final set of otoliths is 9,397 (87.1%).

### Duplicate Check

After QA, we filtered down to a common set of 435 DWPs with 9,258 otoliths and 17,943 hearts and calculated the duplicate rate (% of loci sharing genotypes) for all otoliths and hearts within a DPW for a total of 394,316 total pairwise comparisons.

### Matching Prep

Techniques for matching otoliths back to hearts were refined over time, however, the most definitive work is in this R Notebook `Known sample tests_Kyle.Rmd`, also titled `Matching tissues to otoliths with known samples`, which revisited T028 data to see if `CKMRsim` was better for matching tissues than a standard duplicate check. Andy Barclay had the idea to try `CKMRsim`, and while initial results were promising, they ultimately fared no better than a duplicate check.

Bottom line, this R notebook refined the methods used for otolith-heart matching.

### Previous Matching

Previous matching was completed by Kristen in `old/PWS_pink_match_analysis_all.rmd` with results saved in `../dup_check_results_P066-P067_pass_with_wells.csv`. This is the latest modified file named "dup_check_results*", however, the name of the output file is a bit confusing given that it should have included more than just those two projects. 

However, there are some old R Notebooks that are also titled `old/PWS_pink_match_analysis_P062-P065.Rmd` and `old/PWS_pink_match_analysis_P066-P067.Rmd`. Need to dig into these to see where their output is.

# Import Data

## Genotypes

Using our standard 298 SNP pink salmon GT-seq loci.
```{r}
GCLr::load_objects(path = "../objects", pattern = "loci298")
GCLr::load_objects(path = "../objects", pattern = "LocusControl")
GCLr::load_objects(path = "../objects", pattern = "dup_check_DWP", rds = TRUE)
GCLr::load_objects(path = "../objects", pattern = "DWP_n", rds = TRUE)
GCLr::load_objects(path = "../objects", pattern = "DWPs_postQA", rds = TRUE)
```

Read in postQA `PGOD21.gcl` and heart genotypes for all affected DWPs.
```{r}
GCLr::load_sillys(path = "../data/genotypes/postQA_join_48DWP_dupcheck/", rds = TRUE)
```

# Sample Sizes

How many DWPs, otoliths, and hearts.
```{r}
DWP_n %>% 
  dplyr::filter(`48_DWP_tray_barcode` %in% DWPs_postQA) %>% 
  dplyr::summarise(DWP = dplyr::n(),
                   otolith = sum(otolith),
                   heart = sum(heart))
```

How many pairwise duplicate comparisons?
```{r}
DWP_n %>% 
  dplyr::filter(`48_DWP_tray_barcode` %in% DWPs_postQA) %>% 
  dplyr::mutate(n_comparisons = heart * otolith) %>% 
  dplyr::ungroup() %>% 
  dplyr::summarise(total_comparisons = sum(n_comparisons))
```

# Matching

## Top 2

Get the top 2 matches (ignoring ties!) for each otolith. Should have 9,258 x 2 = 18,516 rows.
```{r}
(
  dup_check_DWP_top2 <-  dup_check_DWP %>%
    dplyr::group_by(otolith_fish_id) %>%
    dplyr::slice_max(duplicate_rate, n = 2, with_ties = FALSE) %>%  # ignoring ties
    dplyr::mutate(diff = max(duplicate_rate) - min(duplicate_rate)) %>%  # difference in % loci matching between best match and 2nd best match
    dplyr::arrange(`48_DWP_tray_barcode`)
)
```

## Top 1

Get the top 1 match (with ties!) for each otolith. Should have 9,258 rows, plus any ties.
```{r}
(
  dup_check_DWP_top1 <-  dup_check_DWP_top2 %>%
    dplyr::slice_max(duplicate_rate, n = 1, with_ties = TRUE) %>%  # best match, with ties
    dplyr::ungroup() %>% 
    dplyr::arrange(`48_DWP_tray_barcode`)
)
```

Quick plot to see what we have.
```{r}
dup_check_DWP_top1 %>% 
  ggplot2::ggplot(ggplot2::aes(x = diff, y = duplicate_rate)) +
  ggplot2::geom_point() + 
  ggplot2::theme_bw()
```

# Visualization

## Transform Data

Need to join in the DWP + position info from PGOD21 for visualization.
```{r}
(
  matchy_data_mod <- dup_check_DWP_top1 %>%
    dplyr::left_join(
      y = PGOD21_DWP.gcl %>% dplyr::select(silly_source, `48_DWP_tray_position`) %>% dplyr::rename(otolith_position = `48_DWP_tray_position`),
      by = dplyr::join_by(otolith_fish_id == silly_source)
    ) %>%  # otolith DWP position: A1, B1, etc.
    dplyr::left_join(
      y = heart_genotypes_DWP.gcl %>% dplyr::select(silly_source, DNA_TRAY_WELL_POS) %>% dplyr::rename(heart_position = DNA_TRAY_WELL_POS),
      by = dplyr::join_by(heart_fish_id == silly_source)
    ) %>%  # heart DWP position: A1, B1, etc.
    tidyr::separate(
      col = otolith_position,
      into = c("otolith_row", "otolith_col"),
      sep = 1,
      remove = FALSE
    ) %>%  # separate row and column designators
    tidyr::separate(
      col = heart_position,
      into = c("heart_row", "heart_col"),
      sep = 1,
      remove = FALSE
    ) %>%
    dplyr::mutate(
      otolith_row = factor(x = otolith_row, levels = LETTERS[8:1]),
      heart_row = factor(x = heart_row, levels = LETTERS[8:1]),
      otolith_col = factor(x = otolith_col, levels = as.character(1:6)),
      heart_col = factor(x = heart_col, levels = as.character(1:6))
    )  # make row and columns factors so they all plot
)
```

## Function

Pull in PGOD21 otolith reads to see if the matches matter if there are close ties?

Any hearts match more than 1 otolith?
```{r}
matchy_data_mod %>%
  dplyr::filter(`48_DWP_tray_barcode` == DWPs_postQA[1])
```


**Red** shows the location of the heart (original location of the otolith, pre-shipment)
**Blue** shows where the otolith ended up at the MTAL (post-shipment)
If there is an arrow, the otolith got shifted from it's original location.
`NA` row and column values for an otolith mean that it was stuck to the lid or between wells
```{r}
matchy_data_mod %>%
  dplyr::filter(`48_DWP_tray_barcode` == DWPs_postQA[1]) %>% {
    ggplot2::ggplot(.) +
      ggplot2::geom_point(
        ggplot2::aes(x = heart_col, y = heart_row, colour = "heart"),
        cex = 8,
        alpha = 0.5
      ) +
      ggplot2::geom_point(
        ggplot2::aes(x = otolith_col, y = otolith_row, colour = "otolith"),
        cex = 6,
        alpha = 0.5
      ) +
      ggplot2::geom_segment(
        data = dplyr::filter(.data = ., heart_row != otolith_row |
                               heart_col != otolith_col),
        ggplot2::aes(
          x = heart_col,
          y = heart_row,
          xend = otolith_col,
          yend = otolith_row
        ),
        arrow = arrow(type = "closed", length = unit(0.5, "cm")),
        lwd = 1.5
      ) +
      ggplot2::scale_x_discrete(drop = FALSE) +  # forces all factor levels to plot
      ggplot2::scale_y_discrete(drop = FALSE) +  # forces all factor levels to plot
      ggplot2::scale_color_manual(
        name = '',
        breaks = c('heart', 'otolith'),
        values = c('heart' = 'red', 'otolith' = 'blue')
      ) +
      ggplot2::xlab("Column") +
      ggplot2::ylab("Row") +
      ggplot2::theme_bw() +
      ggplot2::facet_wrap(facets = ggplot2::vars(`48_DWP_tray_barcode`))
  }
```

Need to consider:
`NA` otolith position
Hearts that best match > 1 otolith and vice versa
Do we want to see otolith count maps?


## Test


# Summary

# Box Otoliths

Don't forget about the 129 "box" otoliths!