---
title: "04 - Matching Hearts & Otoliths"
subtitle: "Assign Otolith Genotypes to Heart Genotypes by DWP"
author: "Kyle Shedd"
date: "Started: 2024-08-15, last opened: `r Sys.Date()`"
output:
  html_notebook:
    theme: united
    toc: yes
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup

Load all necessary packages, as of 2023-08-11 Kyle switched to [GCLr](https://github.com/commfish/GCLr).
```{r setup, message=FALSE, results='hide'}
rm(list = ls(all.names = TRUE))

if(!require("pacman")) install.packages("pacman"); library(pacman)

pacman::p_load(
  tidyverse,
  lubridate,
  scales,
  janitor,
  GCLr
)

knitr::opts_chunk$set(fig.width = 10)

.username = readLines("~/R/usr_pw.txt", n = 1)  # LOKI username
.password = readLines("~/R/usr_pw.txt" , n = 2)[[2]]  # LOKI password
```

# Objective

The purpose of this notebook is to make individual matching decisions for otolith- and heart-genotypes within a deep well plate (DWP) based on the duplicate genotype rate (% of loci sharing genotypes) for the top match and difference between the top and 2nd best match (`diff`).

Following that matching, we will need to consider:  

  1) whether we want to accept any GOD otoliths with valid reads, but no otolith-heart genotype match. For example, some DWPs may not have any GOD otoliths with hatchery marks, in which case heart-otolith pairing is moot. Alternatively, we could have some wells with 2 otoliths and all surrounding wells have been resolved, but no valid heart-otolith match for that well.  
  2) whether any of the PGOD21 otoliths matched up to hearts/fish with "trustworthy", existing, non-GOD otolith reads (i.e. we ended up with a "unknown" [i.e., right-side] otolith in PGOD21, but the left-side otolith stayed put and was already read). Otherwise, we may have duplicate information for the same individual fish.

The next steps after this will be to *verify heart matches are GOD fish (i.e., don't have existing otolith reads)* and *investigate GOD otoliths with reads but no heart match*. 

# Background

During shipment of otolith samples in February 2020 to the MTA in Juneau for reading, a significant proportion of otoliths migrated between cells within DWPs due to poor containment of the acetate lids that were attached with rubber bands to the DWPs. This incident is known as "the great otolith debacle", aka the "GOD" incident. Since some otoliths moved from their original cells in the DWP, the paired integrity of the otolith-origin information and the rest of the paired data (genotype + field data) was lost for just under 11,000 individuals from the following collections:  

  * PERB15
  * PERB16
  * PERB17
  * PGILMOUR15
  * PGILMOUR16
  * PGILMOUR17
  * PGILMOUR18
  * PPADDY16

In an attempt to rectify the "GOD" incident, we determined which otoliths were "suspect" and had likely migrated between wells, transferred left and unknown otoliths to 96 shallow well plates (SWP), extracted DNA from the otolith tissues, genotyped the otolith-derived DNA at 298 GT-seq loci, and are attempting to re-pair the otolith-heart samples from their genotypes. 

In addition to this GitHub repository, other materials can be found on the V: drive here:
[link](V:\Lab\Loki\Collection raw data sheets\Pink\Otolith Extravaganza\GOD Event 2020_2021)

# Methods

Here's a running list of methods and brief summary of results to help me remember how we got here since there are large gaps in time between working on this project.

## Otolith Tranfers

We determined which otoliths were "suspect" and had likely migrated between wells based on comparing the DWP otolith inventory at the MTAL (destination, post-GOD) to the the inventory at the GCL (origin, pre-GOD). Full documentation of the *rules* are located here: `data/raw_working_oto_sheets/working_oto_sheet_combined.xlsm`, tab "decisions". Briefly, any well that had an inventory discrepancy was "bad" and any adjacent well was "suspect" and physically labelled along the left edge of the well.

Only **left or unknown** otoliths from "suspect" wells were transferred to 96 SWPs fro reading (i.e., standard otolith trays) with 1 otolith per well. Full documentation of the *rules* are located in the Excel files here: `data/raw_otolith_transfer_96`, see tab "oto_transfer_business_rules". Right otoliths were also transfered, but not read. After completing the transfers at the MTAL, SWPs were sent to the GCL for DNA extraction and genotyping.

## Laboratory

DNA was extracted from `PGOD21` left-side or unknown otolith tissues of unknown origin using conventional Machery-Nagel DNA extraction kits. Otoliths were placed in T1 buffer for an overnight soak in clear 96 SWPs. Preamp at 14 cycles. Final elution volume was 75uL. All otoliths were transferred among plates with a jig. All liquid handling was done by robot. Samples were genotyped using GT-seq in lab projects P062, P065, P066, P067, P068, P069, P070, P071, P072, P073. QC samples were included in P067, P068, and P074.

Projects P070-072 included an additional step to dry down extracted DNA to increase concentrations and genotyping success, however, this did not occur in P062-P069 and P073. There is a slight difference in average genotyping rate among these projects (75% vs. 68%), but insufficient to warrant re-running P062-P069 and P073. See e-mails from 2024-11-27 for more info.

## Analysis Workflow

### Data Prep

All data prep was done in the R Notebook `01_import-genotypes-join-otlith-transfer-records`. The purpose of that R Notebook was to come up with a definitive link between the `PGOD21.gcl` otoliths and their 96 SWP locations and the *original* 48 DWP or otolith box location. This was accomplished by linking these 3 things:  

  1) `PGOD21.gcl` otolith genotypes from 96 SWP,
  2) Otolith transfer records documenting the transfer of otoliths from their original 48 DWP (or box of 48 DWP),
  3) AHRP Salmon Biological Fact data containing field data (silly code, stream, year, 48 DWP, etc.)
  
Bottom line, we need to know to the finest scale possible (48 DWP or box) where otoliths in `PGOD21.gcl` came from so we can limit our search space for matching.

How many otoliths with genotypes in `PGOD21.gcl`? By collection?

Final stats for `PGOD21.gcl` from `01_import-genotypes-join-otlith-transfer-records`:

 silly_code        n
     PERB15      478
     PERB16    1,527
     PERB17      525
 PGILMOUR15    1,587
 PGILMOUR16    1,555
 PGILMOUR17    2,600
 PGILMOUR18      856
   PPADDY16    1,503
       Box       159
Unaccounted        1
      Total   10,791

match_type         n
          Box    159
 DWP+Position 10,201
     DWP_Only    430
       Nothin      1
        Total 10,791

**439 unique 48 DWPs**

### Quality Assurance (QA)

All QA was done in the R Notebook `02_genotype-QA-hearts-otoliths`. The purpose of that R Notebook was to perform genotyping quality assurance (QA) on the otoliths in PGOD21.gcl and GOD-affected hearts in Erb 2015-2017, Gilmour 2015-2018, and Paddy 2016.

Hearts undergo standard AHRP PWS Pink Salmon QA:  

  1) missing genotypes (<80% loci)
  2) duplicate check (>95% loci, remove both)
  3) contamination (heterozygosity outliers, +/- 3.5 modified z-score, separate for even/odd lineage)

Otoliths undergo a revised QA:  

  1) missing genotypes (<15% loci, arbitrary cutoff)
  2) duplicate check (>99% loci[contamination + low genotyping success], remove 1 [keep sample w/ most location info and genotype data])
  3) contamination (<20 non-heterozygous loci, all together not separate by lineage)

We started with 19,898 hearts from 8 collections. We removed 1,180 (5.9%) due to missing, 60 (0.6%) due to duplicates (both removed), and 592 (3.0%) due to contamination. Final set of hearts is 18,066 (90.8%).

We started with 10,791 otoliths in PGOD21. We removed 522 (4.8%) due to missing, 87 (0.8%) due to duplicates (only 1 removed), and 785 (7.3%) due to contamination. Final set of otoliths is 9,397 (87.1%).

### Duplicate Check

After QA, we filtered down to a common set of 435 DWPs with 9,258 otoliths (85.8%) and 17,943 hearts (90.2%) and calculated the duplicate rate (% of loci sharing genotypes) for all otoliths and hearts within a DPW for a total of 394,316 total pairwise comparisons.

### Matching Prep

Techniques for matching otoliths back to hearts were refined over time, however, the most definitive work is in this R Notebook `Known sample tests_Kyle.Rmd`, also titled `Matching tissues to otoliths with known samples`, which revisited T028 data to see if `CKMRsim` was better for matching tissues than a standard duplicate check. Andy Barclay had the idea to try `CKMRsim`, and while initial results were promising, they ultimately fared no better than a duplicate check.

Bottom line, this R notebook refined the methods used for otolith-heart matching. Using cutoffs for duplicate rate >= 0.8 and diff (best match - 2nd best match) >= 0.25 appears to maximize the number of correct matches while minimizing the number of incorrect matches.

### Previous Matching

Previous matching was completed by Kristen in `old/PWS_pink_match_analysis_all.rmd` with results saved in `../dup_check_results_P066-P067_pass_with_wells.csv`. This is the latest modified file named "dup_check_results*", however, the name of the output file is a bit confusing given that it should have included more than just those two projects. 

However, there are some old R Notebooks that are also titled `old/PWS_pink_match_analysis_P062-P065.Rmd` and `old/PWS_pink_match_analysis_P066-P067.Rmd`. Need to dig into these to see where their output is.

# Import Data

## Genotypes

Using our standard 298 SNP pink salmon GT-seq loci.
```{r}
GCLr::load_objects(path = "../objects", pattern = "loci298")
GCLr::load_objects(path = "../objects", pattern = "LocusControl")
GCLr::load_objects(path = "../objects", pattern = "dup_check_DWP", rds = TRUE)
GCLr::load_objects(path = "../objects", pattern = "DWP_n", rds = TRUE)
GCLr::load_objects(path = "../objects", pattern = "DWPs_postQA", rds = TRUE)
```

Read in postQA `PGOD21.gcl` and heart genotypes for all affected DWPs.
```{r}
GCLr::load_sillys(path = "../data/genotypes/postQA_join_48DWP_dupcheck/", rds = TRUE)
```

## GOD Otolith Reads

Got GOD otolith reads from Cordova Otolith lab (2025-02-05) with help from Eric L. (GCL AP) and Rachel E. (CDV Otolith Lab Manager). Want to see what the reads are for hearts that best match more than one otolith, since they may be from the same fish (left and right).
```{r}
(otolith_reads <- readr::read_csv("../data/GOD_otolith_reads_all.csv") %>% 
   janitor::clean_names())
```

Total of 10,816 otoliths.

## PGOD Paired

Read in all PGOD21 otoliths with paired 48 DWP information. This was generated in `01_import-genotypes-join-otlith-transfer-records.Rmd`. This is what we will use to assign otoliths to hearts. We'll need to add the target (i.e., heart) 48 DWP barcode and DWP position for Eric L. to import into *LOKI*.
```{r}
(PGOD21_48DWP_paired <- readr::read_csv(file = "../output/PGOD21_48DWP_paired.csv"))
```

# Sample Sizes

How many DWPs, otoliths, and hearts.
```{r}
DWP_n %>% 
  dplyr::filter(`48_DWP_tray_barcode` %in% DWPs_postQA) %>% 
  dplyr::summarise(DWP = dplyr::n(),
                   otolith = sum(otolith),
                   heart = sum(heart))
```

How many pairwise duplicate comparisons?
```{r}
DWP_n %>% 
  dplyr::filter(`48_DWP_tray_barcode` %in% DWPs_postQA) %>% 
  dplyr::mutate(n_comparisons = heart * otolith) %>% 
  dplyr::ungroup() %>% 
  dplyr::summarise(total_comparisons = sum(n_comparisons))
```

How many GOD otoliths were successfully read?
```{r}
otolith_reads %>% 
  dplyr::count(mark_present)

otolith_reads %>% 
  dplyr::filter(mark_present == "YES") %>% 
  dplyr::count(mark_id)

otolith_reads %>% 
  dplyr::count(otolith_mark_status_code)
```

9,678 of 10,816 otoliths were successfully read (89.5%). Of those not read, 266 were lost and 760 were unreadable/overground. **NOTE** the otolith codes indicate that almost all are "left", but I'm not sure if that is completely true. Of the 1,648 otoliths with a hatchery mark, most were AFK (unsurpisingly), followed by Cannery Creek, Wally Norenburg, and Solomon Gulch. See <V:/Analysis/5_Coastwide/Multispecies/Alaska Hatchery Research Program/PWS Pink/Otolith Mark Status Codes.xlsx> for an overview of otolith mark status codes.

# Matching

## Output for *LOKI*

Eric Lardizabal ultimately needs to know for each otolith:  

  1) `Target Barcode` = `48_DWP_tray_barcode` (10-digit heart 48DWP barcode)  
  2) `Target Well Position` = `heart_position` (heart 48DWP well position, i.e., A1, A2, etc.)

The two other GOD columns in the *LOKI* tissue table are:  

  * `Determination Collection` = "PGOD21", only used in the target collection (i.e., in PPADDY16 the Determination Collection = PGOD21)  
  * `Target Container Type` = type of container the otolith is going into, should be 96 well SWP

## Top 2

Get the top 2 matches (ignoring ties!) for each otolith. Should have 9,258 x 2 = 18,516 rows.
```{r}
(
  dup_check_DWP_top2 <-  dup_check_DWP %>%
    dplyr::group_by(otolith_fish_id) %>%
    dplyr::slice_max(duplicate_rate, n = 2, with_ties = FALSE) %>%  # ignoring ties
    dplyr::mutate(diff = max(duplicate_rate) - min(duplicate_rate)) %>%  # difference in % loci matching between best match and 2nd best match
    dplyr::arrange(`48_DWP_tray_barcode`)
)
```

## Top 1

Get the top 1 match (with ties!) for each otolith. Should have 9,258 rows, plus any ties.
```{r}
(
  dup_check_DWP_top1 <-  dup_check_DWP_top2 %>%
    dplyr::slice_max(duplicate_rate, n = 1, with_ties = TRUE) %>%  # best match, with ties
    dplyr::ungroup() %>% 
    dplyr::arrange(`48_DWP_tray_barcode`)
)
```

Quick plot to see what we have.
```{r}
dup_check_DWP_top1 %>% 
  ggplot2::ggplot(ggplot2::aes(x = diff, y = duplicate_rate)) +
  ggplot2::geom_point() + 
  ggplot2::theme_bw()
```

## Self - Unmoved Otoliths

Include duplicate rate with self for otoliths that did not move. This is relevant for some uncomplicated DWPs where we genotyped wells that still had 2 otoliths, but were next to wells where otoliths were suspected to have moved (n_otolith != 2).
```{r}
(
  dup_check_DWP_self <- dup_check_DWP %>%
    dplyr::left_join(
      y = PGOD21_DWP.gcl %>% dplyr::select(silly_source, `48_DWP_tray_position`) %>% dplyr::rename(otolith_position = `48_DWP_tray_position`),
      by = dplyr::join_by(otolith_fish_id == silly_source)
    ) %>%  # otolith DWP position: A1, B1, etc.
    dplyr::left_join(
      y = heart_genotypes_DWP.gcl %>% dplyr::select(silly_source, DNA_TRAY_WELL_POS) %>% dplyr::rename(heart_position = DNA_TRAY_WELL_POS),
      by = dplyr::join_by(heart_fish_id == silly_source)
    ) %>%
    dplyr::filter(otolith_position == heart_position) %>%   # filter for otoliths that didn't move
    dplyr::rename(heart_fish_id_self = heart_fish_id,
                  heart_miss_loci_self = heart_miss_loci,
                  duplicate_rate_self = duplicate_rate) %>% 
    dplyr::select(otolith_fish_id, heart_fish_id_self, heart_miss_loci_self, duplicate_rate_self)
)
```

# Visualization

## Join & Transform Data

Need to join in the DWP + position info from PGOD21 for visualization.
```{r}
(
  matchy_data_mod <- dup_check_DWP_top1 %>%
    dplyr::left_join(
      y = PGOD21_DWP.gcl %>% dplyr::select(silly_source, `48_DWP_tray_position`) %>% dplyr::rename(otolith_position = `48_DWP_tray_position`),
      by = dplyr::join_by(otolith_fish_id == silly_source)
    ) %>%  # otolith DWP position: A1, B1, etc.
    dplyr::left_join(
      y = heart_genotypes_DWP.gcl %>% dplyr::select(silly_source, DNA_TRAY_WELL_POS) %>% dplyr::rename(heart_position = DNA_TRAY_WELL_POS),
      by = dplyr::join_by(heart_fish_id == silly_source)
    ) %>%  # heart DWP position: A1, B1, etc.
    tidyr::separate(
      col = otolith_position,
      into = c("otolith_row", "otolith_col"),
      sep = 1,
      remove = FALSE
    ) %>%  # separate row and column designators
    tidyr::separate(
      col = heart_position,
      into = c("heart_row", "heart_col"),
      sep = 1,
      remove = FALSE
    ) %>%
    dplyr::mutate(
      otolith_row = factor(x = otolith_row, levels = LETTERS[8:1]),
      heart_row = factor(x = heart_row, levels = LETTERS[8:1]),
      otolith_col = factor(x = otolith_col, levels = as.character(1:6)),
      heart_col = factor(x = heart_col, levels = as.character(1:6))
    ) %>%   # make row and columns factors so they all plot
    dplyr::left_join(
      y = dup_check_DWP_self,
      by = "otolith_fish_id"
    ) %>% 
    dplyr::mutate(diff_self = duplicate_rate - duplicate_rate_self,
                  match_type = dplyr::case_when(heart_fish_id == heart_fish_id_self ~ "self",
                                                TRUE ~ "not_self")
    ) %>% 
    dplyr::arrange(`48_DWP_tray_barcode`, otolith_fish_id)
)
```

Data is 1 row per otolith, includes best heart match (`top1` could be self or migrated), difference between best and 2nd best match, and self heart match (otolith never moved, matches heart from same well).

## Distribution of Duplicate Rate

Distribution of duplicate rate for top match
```{r}
matchy_data_mod %>% 
  ggplot2::ggplot(ggplot2::aes(x = duplicate_rate, fill = match_type)) +
  ggplot2::geom_histogram(binwidth = .01) + 
  ggplot2::geom_vline(xintercept = 0.8, colour = "red", lwd = 1) +
  ggplot2::labs(title = "Distribution of Locus Duplicate Rate for Best Match for Hearts and Otoliths") +
  ggplot2::theme_bw() +
  ggplot2::facet_grid(rows = dplyr::vars(match_type))
```

The overall distribution is fairly similar to tests with known samples, see histogram in `Known sample tests_Kyle.Rmd`. Kristen and Kyle decided on 2025-01-24 to go with a preliminary cutoff of **> 0.8 duplicate rate**. This resulted in very few incorrect assignments in the known sample tests. We also see that most of the self-assignments (otolith did not migrate to a new well) are above 0.8.

## Distribution of Diff

Distribution of difference rate between top 1 match and 2nd best match.
```{r}
matchy_data_mod %>% 
  ggplot2::ggplot(ggplot2::aes(x = diff, fill = match_type)) +
  ggplot2::geom_histogram(binwidth = .01) + 
  ggplot2::geom_vline(xintercept = 0.25, colour = "red", lwd = 1) +
  ggplot2::labs(title = "Distribution of Difference Between Best and 2nd Best Locus Duplicate Rate Match for Hearts and Otoliths") + 
  ggplot2::theme_bw() +
  ggplot2::facet_grid(rows = dplyr::vars(match_type))
```

The overall distribution is fairly similar to tests with known samples, see histogram in `Known sample tests_Kyle.Rmd`. Kristen and Kyle decided on 2025-01-24 to go with a preliminary cutoff of **> 0.25 diff**. This resulted in very few incorrect assignments in the known sample tests. We also see that most of the self-assignments (otolith did not migrate to a new well) had higher diffs.

## Plot Duplicate Rate vs. Diff

```{r}
matchy_data_mod %>% 
  ggplot2::ggplot(ggplot2::aes(x = diff, y = duplicate_rate, colour = match_type)) +
  ggplot2::geom_point() + 
  ggplot2::theme_bw()
```

## Save Top Duplicate Rate Output

Save output as both .rds and a .csv so Kristen and I can make some matching decisions.
```{r}
GCLr::save_objects(objects = "matchy_data_mod", path = "../objects", rds = TRUE)
readr::write_csv(x = matchy_data_mod, file = "../output/PGOD21_duplicate_rate_top_and_self_matches.csv")
```

## Filter >0.8 Duplicate + >0.25 Diff

Use the preliminary cutoffs that Kristen and I agreed to on 2025-01-24.
```{r}
(
  matchy_data_mod_0.8dup_0.25diff <- matchy_data_mod %>%
    dplyr::filter(duplicate_rate >= 0.8, diff >= 0.25)
)
```

How many self vs. not self?
```{r}
matchy_data_mod_0.8dup_0.25diff %>% 
  dplyr::count(match_type)
```

## Function for DWP Shifting

Create a function for 1 plot per page for each DWP using the filtered cutoffs.
```{r}
matchy_data_mod %>%
  dplyr::filter(`48_DWP_tray_barcode` == DWPs_postQA[1])

matchy_data_mod_0.8dup_0.25diff %>% 
  dplyr::filter(`48_DWP_tray_barcode` == DWPs_postQA[1])
```

**Red** shows the location of the heart (original location of the otolith, pre-shipment)
**Blue** shows where the otolith ended up at the MTAL (post-shipment)
If there is an arrow, the otolith got shifted from it's original location.
`NA` row and column values for an otolith mean that it was stuck to the lid or between wells

**Create 1 plot per DWP in a PDF**

Commented out so as not to re-generate the PDF if need to re-run R Notebook.
These PDFs include multiple-otolith matches for visualization of how otoliths shifted.
```{r}
# pdf(
#   file = "../output/otolith_shift_DWP_visualization_dup-rate-0.8_diff-0.25.pdf",
#   width = 11,
#   height = 8.5,
#   family = "Helvetica",
#   pointsize = 12
# )
# 
# for(DWP in unique(matchy_data_mod_0.8dup_0.25diff$`48_DWP_tray_barcode`)){
# 
# DWP_plot <- matchy_data_mod_0.8dup_0.25diff %>%
#   dplyr::filter(`48_DWP_tray_barcode` == DWP) %>% {
#     ggplot2::ggplot(.) +
#       ggplot2::geom_point(
#         ggplot2::aes(x = heart_col, y = heart_row, colour = "heart"),
#         cex = 8,
#         alpha = 0.5
#       ) +
#       ggplot2::geom_point(
#         ggplot2::aes(x = otolith_col, y = otolith_row, colour = "otolith"),
#         cex = 6,
#         alpha = 0.5
#       ) +
#       ggplot2::geom_segment(
#         data = dplyr::filter(.data = ., heart_row != otolith_row |
#                                heart_col != otolith_col),
#         ggplot2::aes(
#           x = heart_col,
#           y = heart_row,
#           xend = otolith_col,
#           yend = otolith_row
#         ),
#         arrow = arrow(type = "closed", length = unit(0.5, "cm")),
#         lwd = 1.5
#       ) +
#       ggplot2::scale_x_discrete(drop = FALSE) +  # forces all factor levels to plot
#       ggplot2::scale_y_discrete(drop = FALSE) +  # forces all factor levels to plot
#       ggplot2::scale_color_manual(
#         name = '',
#         breaks = c('heart', 'otolith'),
#         values = c('heart' = 'red', 'otolith' = 'blue')
#       ) +
#       ggplot2::xlab("Column") +
#       ggplot2::ylab("Row") +
#       ggplot2::theme_bw() +
#       ggplot2::facet_wrap(facets = ggplot2::vars(`48_DWP_tray_barcode`))
#   }
#     
#   grid::grid.newpage()
#   grid::grid.draw(rbind(
#     ggplot2::ggplotGrob(DWP_plot),
#     size = "last"
#   ))
#   
# }
# 
# dev.off()
```

# Multiple Matches?

## No Filters

### Otolith Matches >1 Heart

Any otoliths best match more than one heart?
```{r}
matchy_data_mod %>% 
  dplyr::group_by(otolith_fish_id) %>% 
  dplyr::filter(dplyr::n() > 1) %>% 
  dplyr::arrange(dplyr::desc(duplicate_rate))
```

Yes, but duplicate rates are all pretty low, < 0.67. These are likely bogus matches.

### Heart Matches >1 Otolith

Any hearts best match more than one otolith?
```{r}
matchy_data_mod %>% 
  dplyr::group_by(heart_fish_id) %>% 
  dplyr::filter(dplyr::n() > 1) %>% 
  dplyr::arrange(heart_fish_id)

matchy_data_mod %>% 
  dplyr::group_by(heart_fish_id) %>% 
  dplyr::filter(dplyr::n() > 1) %>% 
  ggplot2::ggplot(ggplot2::aes(x = heart_fish_id, y = duplicate_rate, colour = match_type)) +
  ggplot2::geom_point() + 
  ggplot2::theme_bw()
```

Yes, there are several hearts that are the best match for > 1 otolith, however, most of them most closely match themselves. It's okay for a heart to be the best match for up to 2 otoliths (left and right), so long as those otoliths have the same read (hatchery vs. wild + origin).

## Filter >0.8 Duplicate + >0.25 Diff

### Heart Matches >1 Otolith

Any hearts best match more than one otolith now that we've applied our `duplicate_rate` and `diff` cutoffs?
```{r}
matchy_data_mod_0.8dup_0.25diff %>% 
  dplyr::group_by(heart_fish_id) %>% 
  dplyr::filter(dplyr::n() > 1) %>% 
  dplyr::arrange(heart_fish_id)

matchy_data_mod_0.8dup_0.25diff %>% 
  dplyr::count(heart_fish_id) %>% 
  dplyr::count(n, name = "n_matching_otos_per_heart")
```

Yes, there are 296 hearts that best match 2 otoliths (may be lefts and unknown [rights]?) and 10 hearts that best match 3 otoliths (pieces? or bogus matches).

#### Join GOD Otolith Read Data

Need to bring in the otolith reads to see if this matters or not. If both (or all 3) matching otoliths have the same read (hatchery/wild)...no big deal.
```{r}
(
  matchy_data_mod_0.8dup_0.25diff_multi_oto_match <- matchy_data_mod_0.8dup_0.25diff %>%
    dplyr::group_by(heart_fish_id) %>%
    dplyr::filter(dplyr::n() > 1) %>%  # filter for multiple matches
    dplyr::arrange(heart_fish_id) %>%
    dplyr::left_join(
      y = PGOD21_DWP.gcl %>% dplyr::select(silly_source, oto_tray_and_well),
      by = dplyr::join_by(otolith_fish_id == silly_source)
    ) %>%  # get 96 otolith tray barcode and well for joining to otolith reads
    dplyr::left_join(
      y = otolith_reads %>%
        dplyr::mutate(
          oto_tray = as.numeric(tray_id),
          oto_well = as.numeric(cell_number)
        ) %>%
        tidyr::unite(col = "oto_tray_and_well", c(oto_tray, oto_well), sep = "_") %>%
        dplyr::select(oto_tray_and_well, mark_present, mark_id, otolith_mark_status_code),
      by = "oto_tray_and_well"
    ) # join otolith reads
)
```

#### Conflicts

Hearts with `mark_present` conflicts.
```{r}
# only looking at the mark present/absent level, not at the individual hatchery_id level yet
heart_fish_id_match_multi_oto_conflict <- matchy_data_mod_0.8dup_0.25diff_multi_oto_match %>% 
  dplyr::filter(!is.na(mark_present)) %>%  # remove NA otolith reads
  dplyr::count(heart_fish_id, mark_present) %>%
  dplyr::count(heart_fish_id) %>% 
  dplyr::filter(n > 1) %>%  # filter for hearts that match both a hatchery and wild otolith read
  dplyr::pull(heart_fish_id)  # 11 hearts

matchy_data_mod_0.8dup_0.25diff_multi_oto_match %>% 
  dplyr::filter(heart_fish_id %in% heart_fish_id_match_multi_oto_conflict)
```

There are 11 hearts that match > 1 otolith with different hatchery vs. wild otolith reads. While some of these have very high duplicate rates or there is a big disparity in the number of missing loci for the otolith genotype, it might just be cleanest to drop all 11 hearts.

Hearts with `mark_id` conflicts.
```{r}
# looking at the individual hatchery_id level
heart_fish_id_match_multi_oto_conflict_mark_id <- matchy_data_mod_0.8dup_0.25diff_multi_oto_match %>% 
  dplyr::filter(mark_present == "YES") %>%  # remove NA otolith reads
  dplyr::count(heart_fish_id, mark_id) %>%
  dplyr::count(heart_fish_id) %>% 
  dplyr::filter(n > 1) %>%  # filter for hearts that match both a hatchery and wild otolith read
  dplyr::pull(heart_fish_id)  # 11 hearts

matchy_data_mod_0.8dup_0.25diff_multi_oto_match %>% 
  dplyr::filter(heart_fish_id %in% heart_fish_id_match_multi_oto_conflict_mark_id)
```

Looks like there are 3 hearts that closely match 2 otoliths each, but clearly from different fish due to different otolith reads. Two are vial otoliths. Cleanest to drop all 3 hearts.

Hearts with `mark_present` conflicts including `NA` otolith reads.
```{r}
heart_fish_id_match_multi_oto_conflict_mark_present_NA <- matchy_data_mod_0.8dup_0.25diff_multi_oto_match %>% 
  dplyr::filter(!heart_fish_id %in% heart_fish_id_match_multi_oto_conflict) %>% 
  dplyr::filter(!heart_fish_id %in% heart_fish_id_match_multi_oto_conflict_mark_id) %>% 
  dplyr::count(heart_fish_id, mark_present) %>%
  dplyr::count(heart_fish_id) %>% 
  dplyr::filter(n > 1) %>%  # filter for hearts with different mark_present (i.e., NA and NO, or NA and YES)
  dplyr::pull(heart_fish_id)  # 66 hearts

matchy_data_mod_0.8dup_0.25diff_multi_oto_match %>% 
  dplyr::filter(heart_fish_id %in% heart_fish_id_match_multi_oto_conflict_mark_present_NA)
```

#### Keep

For hearts with two matches, and the same otolith read (`mark_present` and `mark_id`), pick the highest duplicate rate to keep for assignment.
```{r}
otolith_fish_id_match_multi_oto_keep <- matchy_data_mod_0.8dup_0.25diff_multi_oto_match %>% 
  dplyr::filter(!heart_fish_id %in% heart_fish_id_match_multi_oto_conflict) %>%  # drop hearts with conflicting `mark_present` (i.e., NO vs. YES)
  dplyr::filter(!heart_fish_id %in% heart_fish_id_match_multi_oto_conflict_mark_id) %>%  # drop hearts with conflicting `mark_id`
  dplyr::filter(!heart_fish_id %in% heart_fish_id_match_multi_oto_conflict_mark_present_NA) %>%  # drop hearts with conflicting NA `mark_present` (i.e., NA vs. NO or NA vs. YES)
  dplyr::slice_max(duplicate_rate, n = 1, with_ties = FALSE) %>%  # if same otolith read, pick highest duplicate rate
  dplyr::ungroup() %>% 
  dplyr::pull(otolith_fish_id)

matchy_data_mod_0.8dup_0.25diff_multi_oto_match %>% 
  dplyr::filter(otolith_fish_id %in% otolith_fish_id_match_multi_oto_keep)
```

Alternate method for a sanity check. Should get 226 heart-otolith matches with the same otolith reads.
```{r}
matchy_data_mod_0.8dup_0.25diff_multi_oto_match %>% 
  dplyr::count(heart_fish_id, mark_present, mark_id) %>% 
  dplyr::count(heart_fish_id) %>% 
  dplyr::filter(n == 1)
```

## Finalize

Subset matches to toss multiple matches with conflicts. These are the **final** matches for this notebook. In a future notebook, we'll still need to:  

  1) make sure that none of the `heart_fish_id` already have existing otolith reads (i.e., did we assign a GOD otolith to a non-GOD heart?), and 
  2) potentially add some additional assignments if all remaining GOD otoliths have the same read
```{r}
(
  matchy_data_mod_0.8dup_0.25diff_no_multi_oto_match <-
    matchy_data_mod_0.8dup_0.25diff %>%
    dplyr::filter(
      !otolith_fish_id %in% setdiff(
        matchy_data_mod_0.8dup_0.25diff_multi_oto_match$otolith_fish_id,
        otolith_fish_id_match_multi_oto_keep
      )
    )
)
```

# Assign Matches

Assign PGOD21 otoliths to hearts with matches with `duplicate_rate >= 0.8` and `diff >= 0.25` thresholds. If multiple otoliths match the same heart, pick the best match (highest duplicate rate) if otoliths have the same read, otherwise remove both.
```{r}
PGOD21_DWP_passQA.gcl <- readRDS(file = "../data/genotypes/postQA_join_48DWP/PGOD21.rds")  # pull in fish that passed QA in 02_genotype-QA-hearts-otoliths

(
  PGOD21_48DWP_paired_genotype_matches <- PGOD21_48DWP_paired %>%
    dplyr::mutate(
      otolith_info = dplyr::case_when(
        !is.na(`48_DWP_tray_barcode`) & !is.na(`48_DWP_tray_position`) ~ "DWP+Position",
        !is.na(`48_DWP_tray_barcode`) & is.na(`48_DWP_tray_position`) ~ "DWP_Only",
        is.na(`48_DWP_tray_barcode`) & !is.na(box_number) ~ "Box",
        is.na(`48_DWP_tray_barcode`) & is.na(box_number) ~ "Nothin",
        TRUE ~ "WTF"
      ),  # define match_type from 01_import-genotypes-join-otlith-transfer-records
      pass_QA = SillySource %in% PGOD21_DWP_passQA.gcl$SillySource  # which fish passed QA?
    ) %>%
    dplyr::left_join(
      y = matchy_data_mod_0.8dup_0.25diff_no_multi_oto_match %>%
        dplyr::select(
          otolith_fish_id,
          heart_fish_id,
          `48_DWP_tray_barcode`,
          heart_position
        ),
      by = dplyr::join_by("SillySource" == "otolith_fish_id", `48_DWP_tray_barcode`)
    ) %>%   # join in final duplicate matches after dealing with hearts that matches > 1 otolith
    dplyr::mutate(genotype_match_assigned = !is.na(heart_fish_id),
                  otolith_moved = dplyr::case_when(
                    genotype_match_assigned & `48_DWP_tray_position` == heart_position ~ FALSE,
                    genotype_match_assigned & `48_DWP_tray_position` != heart_position ~ TRUE,
                    !genotype_match_assigned ~ NA)) %>% 
    dplyr::relocate(genotype_match_assigned:otolith_moved, .after = pass_QA)
)

readr::write_csv(x = PGOD21_48DWP_paired_genotype_matches, file = "../output/PGOD21_48DWP_paired_genotype_matches.csv")
```

How'd we do? How many PGOD21 otoliths had a solid match?
```{r}
PGOD21_48DWP_paired_genotype_matches %>% 
  dplyr::count(genotype_match_assigned) %>% 
  dplyr::mutate(p = round(n / sum(n), 3))

PGOD21_48DWP_paired_genotype_matches %>% 
  dplyr::count(otolith_info, genotype_match_assigned) %>% 
  tidyr::pivot_wider(names_from = genotype_match_assigned, values_from = n)
```

Just over 50% of PGOD21 otoliths were assigned. Have not assigned any box otoliths, going to drop them...

Of the assigned otoliths and had DWP+Position, how many moved?
```{r}
PGOD21_48DWP_paired_genotype_matches %>% 
  dplyr::filter(genotype_match_assigned == TRUE) %>% 
  dplyr::count(otolith_moved) %>% 
  dplyr::mutate(p = round(n / sum(n), 3))
```

# Future Work

The next steps after this will be to *verify heart matches are GOD fish (i.e., don't have existing otolith reads)* and *investigate GOD otoliths with reads but no heart match*.

# Summary

We took the within DWP duplicate checks for 9,258 post-QA otoliths to determine the highest `duplicate_rate`  and `diff` between duplicate rates for the 1st best and 2nd best match. Based on results from T028 in `Known sample tests_Kyle.Rmd`, Kristen and I decided to use threshold cutoffs of `duplicate_rate >= 0.8` and `diff >= 0.25` for assigning matches, as this minimized false positives. There is a .pdf in `output` with visualizations of how otoliths shifted within 48 DWPs, which can be compared to the original DWP otolith inventories `raw_working_oto_sheets`. No otoliths above this threshold matched > 1 heart (i.e., no ties). There were 306 hearts that were matched to 2 or 3 otoliths above threshold cutoffs. Of those, we kept 226 hearts that matched to otoliths with the same otolith reads (identical `mark_present` and `mark_id`) and the remaining matches were dropped.

End