---
title: "Hogan 2017/2019 Exploratory Analysis with 298 Markers"
output:
  html_notebook:
    theme: united
    toc: yes
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
rm(list=ls())

knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(lubridate)
library(DT)
library(abind)
bbind <- function(...) { abind(..., along = 3) }

# mod for user
.username <- "kmgruenthal"
.password <- ""
source("~/R/Functions.GCL.R")
```

# Introduction

This is a quick exploratory analysis of the Hogan 2017/2019 genotype data. Here we will:

  * Read in *LocusControl*
  * Read in all project genotypes
  * Read in paired field/otolith data from [OceanAK](http://www.oceanak.adfg.alaska.gov/)
  * Join genetic data with paired data from OceanAK
    + Will need to create data key for join (DWP barcode + "_" + DWP position number)
  * Determine *FailureRate*
  * Perform QA on project gentoypes
    + Removing fish missing >= 20% of genotypes
    + Remove within-collection duplicates
    + Use a heterozygosity filter to remove individuals with bogus (contaminated) genotype calls
  * Save final, post-QA genotypes with paired data
  * Export genotypes as a *Genepop* file to feed in to *genepop2franz.GCL.r*
  
# Read in genotypes
Create *LocusControl* and read in genotypes by *silly*
```{r loki}
sillyvec <- ProjectSillys <- c("PHOGAN17", "PHOGAN19")
loci298 <- dget("../../Objects/loci298.txt")

read_project_genotypes.GCL(sillyvec = sillyvec, loci = loci298, username = .username, password = .password) # Update 11/26/19, LOKI2R worked after Eric contacted IT, but we are retaining read_project_genotypes for now as LOKI2R pulls in extraneous info we do not need

loci <- LocusControl$locusnames
nalleles <- LocusControl$nalleles
ploidy <- LocusControl$ploidy
alleles <- LocusControl$alleles

rm(.username, .password)
```

Save objects
```{r save_locus_control_sillys}
if(!dir.exists("../../Objects/Hogan_17_19/")) {dir.create("../../Objects/Hogan_17_19/")}
save_objects(objects = c("LocusControl", "loci", "sillyvec", "ProjectSillys"), path = "../../Objects/Hogan_17_19/")

if(!dir.exists("../../Genotypes/Hogan_17_19/")) {dir.create("../../Genotypes/Hogan_17_19/")}
save_sillys(sillyvec = sillyvec, path = "../../Genotypes/Hogan_17_19/")
```

# OceanAK

## Read data
Here, we read in paired field and otolith data from *OceanAK* and join that data to the attributes table of each silly object.
```{r read_oceanak, message=FALSE}
# The `genepop2franz.GCL` function we use in the very last step expects the *OceanAK* file to have the "old" names (lowercase with spaces) and not the "new" names (UPPERCASE with _); we convert the column names to the old format here. Note that this is a patch and column names should be double-checked during each new analysis.
og_names <- suppressMessages(names(read_csv(file = "../../OceanAK/PedigreeData_AHRP - Salmon Biological Data 2_PWS_2013-2018_no_otoliths.csv", progress = FALSE)))
oceanak <- read_csv(file = "../../OceanAK/AHRP Salmon Biological Data 20210115_100431.csv") # recommend data pull from OceanAK prior to running
names(oceanak) <- og_names
write_csv(oceanak, "../../OceanAK/AHRP Salmon Biological Data 20210115_100431_names.csv")

oceanak %>% 
  count(`Sample Year`, `Location Code`) %>% 
  spread(`Sample Year`, n, fill = 0)
```

Subset
```{r}
hogan_oceanak <- oceanak %>% 
  dplyr::filter(`Location Code` == "Hogan Creek" & `Sample Year` %in% c(2017, 2019)) %>% 
  tidyr::unite(silly_source, c("Silly Code", "Fish ID"), sep = "_", remove = FALSE)
```

```{r}
hogan_oceanak %>% 
  count(`Sample Year`, `Otolith Mark Present`) %>% 
  spread(`Otolith Mark Present`, n, fill = 0)
```

Verify all genotypes in OceanAK
```{r}
table(PHOGAN17.gcl$attributes$SillySource %in% hogan_oceanak$silly_source)
table(PHOGAN19.gcl$attributes$SillySource %in% hogan_oceanak$silly_source)
```

What attribute names do we have since we used `read_project_genotypes.GCL` rather than `LOKI2R.GCL`?
```{r}
names(PHOGAN17.gcl$attributes)
```

## Join attributes table
Join *hogan_oceanak* data to the attributes table of each silly. Already looked to verify that no tissues were missing or duplicated, which makes sense because the lab can't genotype tissue that aren't there and they never genotype two samples in a well.
```{r join_oceanak_hogan}
PHOGAN17.gcl$attributes <- PHOGAN17.gcl$attributes %>% 
  dplyr::left_join(hogan_oceanak, by = c("SillySource" = "silly_source")) %>% 
  dplyr::select(FK_FISH_ID, SILLY_CODE, PLATE_ID, `DNA Tray Code`, `DNA Tray Well Code`, SillySource, `Sample Date`, Sex, `Length Mm`, `Location Code`, `Otolith Mark Present`, `Otolith Mark ID`, `Sample Year`) %>% 
  dplyr::rename(DNA_TRAY_CODE = "DNA Tray Code",
                DNA_TRAY_WELL_CODE = "DNA Tray Well Code")

PHOGAN19.gcl$attributes <- PHOGAN19.gcl$attributes %>% 
  dplyr::left_join(hogan_oceanak, by = c("SillySource" = "silly_source")) %>% 
  dplyr::select(FK_FISH_ID, SILLY_CODE, PLATE_ID, `DNA Tray Code`, `DNA Tray Well Code`, SillySource, `Sample Date`, Sex, `Length Mm`, `Location Code`, `Otolith Mark Present`, `Otolith Mark ID`, `Sample Year`) %>% 
  dplyr::rename(DNA_TRAY_CODE = "DNA Tray Code",
                DNA_TRAY_WELL_CODE = "DNA Tray Well Code")

if(!dir.exists("../../Genotypes/Hogan_17_19/OceanAK")) {dir.create("../../Genotypes/Hogan_17_19/OceanAK")}
save_sillys(sillyvec = sillyvec, path = "../../Genotypes/Hogan_17_19/OceanAK")
```

## Split by otolith-origin
First we need to verify that we only have hatchery and natural-origin fish for the parental year (2017), and only natural-origin fish for the offspring year (2019).
```{r view_origin}
dplyr::bind_rows(PHOGAN17.gcl$attributes, 
                 PHOGAN19.gcl$attributes) %>% 
  count(SILLY_CODE, `Otolith Mark Present`) %>% 
  spread(`Otolith Mark Present`, n, fill = 0)
```
There are zero NAs.

We want to remove any NA from 2017 and hatchery-origin fish + NA from 2019. We'll need those hatchery-origin fish from 2017 for looking at 2017/2019 parentage. The NAs must be from second reads.
```{r remove_hogan_hatchery}
PHOGAN17_ids_remove <- PHOGAN17.gcl$attributes %>%
  filter(is.na(`Otolith Mark Present`)) %>%
  mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>%
  pull(FK_FISH_ID)
RemoveIDs.GCL(silly = "PHOGAN17", IDs = PHOGAN17_ids_remove)

PHOGAN19_ids_remove <- PHOGAN19.gcl$attributes %>% 
  filter(`Otolith Mark Present` == "YES" | is.na(`Otolith Mark Present`)) %>% 
  mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
  pull(FK_FISH_ID)
RemoveIDs.GCL(silly = "PHOGAN19", IDs = PHOGAN19_ids_remove)
```

Verify
```{r verify_origin}
dplyr::bind_rows(PHOGAN17.gcl$attributes, 
                 PHOGAN19.gcl$attributes) %>% 
  count(SILLY_CODE, `Otolith Mark Present`) %>% 
  spread(`Otolith Mark Present`, n, fill = 0)
```

Now that we have removed those fish that we did not want due to their otolith status, we will split 2017 otolith-origin for QA tabulating purposes. Also create a new *sillyvec*.
```{r split_hogan15}
PHOGAN17_ids_natural <- list("PHOGAN17" = PHOGAN17.gcl$attributes %>% 
  filter(`Otolith Mark Present` == "NO") %>% 
  mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
  pull(FK_FISH_ID))
PoolCollections.GCL(collections = "PHOGAN17", loci = loci, IDs = PHOGAN17_ids_natural, newname = "PHOGAN17n")

PHOGAN17_ids_hatchery <- list("PHOGAN17" = PHOGAN17.gcl$attributes %>% 
  filter(`Otolith Mark Present` == "YES") %>% 
  mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
  pull(FK_FISH_ID))
PoolCollections.GCL(collections = "PHOGAN17", loci = loci, IDs = PHOGAN17_ids_hatchery, newname = "PHOGAN17h")

ProjectSillys_Origin <- c("PHOGAN17n", "PHOGAN17h", "PHOGAN19")
if(!dir.exists("../../Genotypes/Hogan_17_19/OceanAK_Origin")) {dir.create("../../Genotypes/Hogan_17_19/OceanAK_Origin")}
save_sillys(ProjectSillys_Origin, "../../Genotypes/Hogan_17_19/OceanAK_Origin/")
```

# Failure rate
Calculate and view the failure rate by silly, locus, and plate.
```{r failure_rate}
project <- "Hogan_17_19" #this is required of the function; can be any name 
failure_rate <- FailureRate.GCL(sillyvec = ProjectSillys_Origin)
failure_rate_noplots <- failure_rate[1:4]
save_objects("failure_rate_noplots", "../../Objects/Hogan_17_19")
failure_rate
```
There's a 5% genotype failure rate in PHOGAN17 and 7% in PHOGAN19, with an overall rate of 6%. There are some specific plate effects, but they are very scattered, as are any locus-specific failures. 

# QA
Go through our standard Quality Assurance (QA) process, explore heterozygosity, and produce a matrix with sample sizes.
```{r qa_setup}
ProjectSillys_Origin_SampleSizes <- matrix(data = NA, nrow = length(ProjectSillys_Origin), ncol = 5, dimnames = list(ProjectSillys_Origin, c("Genotyped", "Missing", "Duplicate", "Heterozygosity", "Final")))
ProjectSillys_Origin_SampleSizes[, "Genotyped"] <- sapply(paste0(ProjectSillys_Origin, ".gcl"), function(x) get(x)$n)
```

## Missing
Remove individuals missing >=20% of genotypes (i.e. the 80% rule).
```{r qa_missing}
MissLoci <- RemoveIndMissLoci.GCL(sillyvec = ProjectSillys_Origin, proportion = 0.8) 
MissLoci  # character vector of fish removed
save_objects("MissLoci", path = "../../Objects/Hogan_17_19/")
ProjectSillys_Origin_SampleSizes[, "Missing"] <-  ProjectSillys_Origin_SampleSizes[, "Genotyped"] - sapply(paste0(ProjectSillys_Origin, ".gcl"), function(x) get(x)$n)
```
Removed 210 fish IDs from PHOGAN17n, 65 from PHOGAN17h, and 321 from PHOGAN19.

## Duplicates
Remove duplicate individuals within the same collection. Typically, we specify *duplicates* as a pair of individuals that share >=95% of genotypes. Once a pair of *duplicates* is identified, we keep the individual with the most genotypes and remove the other(s).
```{r qa_duplicate}
DuplicateCheck95MinProportion <- CheckDupWithinSilly.GCL(sillyvec = ProjectSillys_Origin, loci = loci, quantile = NULL, minproportion = 0.95)
DuplicateCheckReportSummary <- sapply(ProjectSillys_Origin, function(x) DuplicateCheck95MinProportion[[x]]$report, simplify = FALSE)
DuplicateCheckReportSummary
save_objects("DuplicateCheckReportSummary", path = "../../Objects/Hogan_17_19/")
```
There were two apparent duplicates in PHOGAN17n and one in PHOGAN19. There were no duplicates in PHOGAN17h.

```{r remove all duplicates}
#RemovedDups <- RemoveDups.GCL(DuplicateCheck95MinProportion)
#save_objects("RemovedDups", path = "../Objects/")

# All duplicate IDs in to one tibble
dups <- bind_rows(DuplicateCheckReportSummary[sapply(DuplicateCheckReportSummary, is.data.frame)], .id = "silly") %>% 
  select(silly, ID1, ID2) %>% 
  gather(trash, id, -silly) %>% 
  select(-trash) %>% 
  arrange(silly, id)

# Separate vector of IDs for each silly
dup_17n_ids <- filter(dups, silly == "PHOGAN17n") %>% pull(id)
# dup_17h_ids <- filter(dups, silly == "PHOGAN17h") %>% pull(id)
dup_19_ids <- filter(dups, silly == "PHOGAN19") %>% pull(id)

# Save attributes for posterity
# ID silly sources for duplicate individuals
dup_17n <- PHOGAN17n.gcl$attributes %>% 
  filter(FK_FISH_ID %in% dup_17n_ids) %>% 
  mutate(FK_FISH_ID = as.numeric(FK_FISH_ID))
# dup_17h <- PHOGAN17h.gcl$attributes %>% 
#   filter(FK_FISH_ID %in% dup_17h_ids) %>% 
#   mutate(FK_FISH_ID = as.numeric(FK_FISH_ID))
dup19 <- PHOGAN19.gcl$attributes %>%  
  filter(FK_FISH_ID %in% dup_19_ids)

# One tibble with all attributes for fish to remove
duplicates_to_remove <- bind_rows(dup_17n, dup19)
save_objects("duplicates_to_remove", path = "../../Objects/Hogan_17_19/")

# Remove fish from .gcl objects
RemoveIDs.GCL(silly = "PHOGAN17n", IDs = dup_17n_ids)
# RemoveIDs.GCL(silly = "PHOGAN17h", IDs = dup_17h_ids)
RemoveIDs.GCL(silly = "PHOGAN19", IDs = dup_19_ids)

# Fill in sample size matrix
ProjectSillys_Origin_SampleSizes[, "Duplicate"] <-  ProjectSillys_Origin_SampleSizes[, "Genotyped"] - ProjectSillys_Origin_SampleSizes[, "Missing"] - sapply(paste0(ProjectSillys_Origin, ".gcl"), function(x) get(x)$n)
```

## Heterozygosity
We know from the QC work that the **GTscore** genotyping pipeline can sometimes result in individuals with an overabundance of heterozygous calls (perhaps from contamination). We want to remove individuals with untrustworthy genotypes. To explore the utility of a *heterozygosity filter*, we should first plot a histogram of heterozygosity per individual to see what the problem looks like. From there we can set a threshold or range of acceptable heterozygosities.
```{r heterozygosity_calculations}
# genotype.counts <- apply(PHOGAN13.gcl$scores, MARGIN=1, table) # provides counts of each genotype/individual

# Calculate heterozygosity by individual :-) 
PHOGAN17n_ind_het <- apply(PHOGAN17n.gcl$scores, 1, function(ind) {sum(ind[ ,"Dose1"] != ind[ ,"Dose2"])/298} ) 
PHOGAN17h_ind_het <- apply(PHOGAN17h.gcl$scores, 1, function(ind) {sum(ind[ ,"Dose1"] != ind[ ,"Dose2"])/298} ) 
PHOGAN19_ind_het <- apply(PHOGAN19.gcl$scores, 1, function(ind) {sum(ind[ ,"Dose1"] != ind[ ,"Dose2"])/298} )

# Convert to tibbles
tibble_PHOGAN17n <- tibble(silly=PHOGAN17n.gcl$attributes$SILLY_CODE, sillysource=PHOGAN17n.gcl$attributes$SillySource, het=PHOGAN17n_ind_het)
tibble_PHOGAN17h <- tibble(silly=PHOGAN17h.gcl$attributes$SILLY_CODE, sillysource=PHOGAN17h.gcl$attributes$SillySource, het=PHOGAN17h_ind_het)
tibble_PHOGAN19 <- tibble(silly=PHOGAN19.gcl$attributes$SILLY_CODE, sillysource=PHOGAN19.gcl$attributes$SillySource, het=PHOGAN19_ind_het)

# Create a single tibble
het_tibble <- bind_rows(tibble_PHOGAN17n, tibble_PHOGAN17h, tibble_PHOGAN19)
```

### Plot by lineage and remove outliers
Plot heterozygosities with *tidyverse* for ease of use/visualization. Cutoffs (red vertical lines) are based on the +/- 1.5 IQR method that it is well known and well documented in literature.
```{r visualize heterozygosity by lineage}
# NOTE: Original plotting by silly removed and +/- 1.5 IQR heterozygosity cutoff instilled -- After discussions among Chris, Tyler, Dan, and Kyle, they decided it would be more appropriate to examine heterozygosity by lineage (even and/or odd) instead of year. It was also agreed that the +/- 3SD cutoff (Anderson et al. 2010) was not restrictive enough as the distribution has a long right-sided tail and the upper bound for the cutoff should be closer to the main distribution. A hybrid righ-side mirror approach based on the +/- 3SD was also suggested by Chris and discarded in favor of the +/- 1.5 IQR.

# combine tibbles and add lineage as grouping variable
het_tibble <- het_tibble %>% 
  dplyr::mutate(lineage = dplyr::case_when(silly %in% c("PHOGAN17", "PHOGAN19") ~ "odd",
                           TRUE ~ "odd"))
save_objects("het_tibble", "../../Objects/Hogan_17_19/")

mean_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(het=mean(het))

low_box_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(het = quantile(het, probs = 0.25) - 1.5 * diff(quantile(het, probs = c(0.25, 0.75))))

high_box_het <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(het = quantile(het, probs = 0.75) + 1.5 * diff(quantile(het, probs = c(0.25, 0.75))))

het_tibble %>% 
  ggplot2::ggplot(aes(x = het)) +
  ggplot2::geom_histogram(binwidth = 1/298) +
  ggplot2::geom_vline(aes(xintercept = het, color = "mean"), mean_het) +
  ggplot2::geom_vline(aes(xintercept = het, color = "IQR"), low_box_het) +
  ggplot2::geom_vline(aes(xintercept = het, color = "IQR"), high_box_het) +
  ggplot2::facet_grid(rows=vars(lineage), scales = "free_y") +
  ggplot2::scale_color_manual(name="statistics", values=c(mean="blue", IQR="red")) +
  ggplot2::xlim(0.15, 0.85) +
  ggplot2::xlab("Individual Heterozygosity") +
  ggplot2::ylab("Frequency") +
  ggplot2::ggtitle("Individual Heterozygosity By Lineage")

```

### Remove fish outside of the IQR cutoffs
How many fish are we going to remove from each silly?
```{r table_het_cutoff}
# Calculate 1.5 IQR
het_IQR_cutoffs <- het_tibble %>% 
  dplyr::group_by(lineage) %>% 
  dplyr::summarise(low_1.5_IQR = quantile(het, probs = 0.25) - 1.5 * diff(quantile(het, probs = c(0.25, 0.75))),
            high_1.5_IQR = quantile(het, probs = 0.75) + 1.5 * diff(quantile(het, probs = c(0.25, 0.75))))

# Get list of fish to remove
het_tibble %>% 
  dplyr::left_join(het_IQR_cutoffs, by = "lineage") %>% 
  dplyr::filter(het < low_1.5_IQR | het > high_1.5_IQR) %>% 
  dplyr::count(silly)
```
We'll remove another 108 fish total.

To remove these fish, we first need to identify them and convert *SillySource* to *FK_FISH_ID* (**note**: the values for *FK_FISH_ID* got reassigned after pooling) before removal.
```{r remove_het_cutoff}
removed_hets <- het_tibble %>% 
  dplyr::left_join(het_IQR_cutoffs, by = "lineage") %>% 
  dplyr::filter(het < low_1.5_IQR | het > high_1.5_IQR) %>% 
  dplyr::pull(sillysource)
save_objects("removed_hets", "../../Objects/Hogan_17_19/")

PHOGAN17h_hets_remove <- PHOGAN17h.gcl$attributes %>% 
  dplyr::filter(SillySource %in% removed_hets) %>% 
  dplyr::mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
  dplyr::pull(FK_FISH_ID)
RemoveIDs.GCL(silly = "PHOGAN17h", IDs = PHOGAN17h_hets_remove)

PHOGAN17n_hets_remove <- PHOGAN17n.gcl$attributes %>% 
  dplyr::filter(SillySource %in% removed_hets) %>% 
  dplyr::mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
  dplyr::pull(FK_FISH_ID)
RemoveIDs.GCL(silly = "PHOGAN17n", IDs = PHOGAN17n_hets_remove)

PHOGAN19_hets_remove <- PHOGAN19.gcl$attributes %>% 
  dplyr::filter(SillySource %in% removed_hets) %>% 
  dplyr::mutate(FK_FISH_ID = as.character(FK_FISH_ID)) %>% 
  dplyr::pull(FK_FISH_ID)
RemoveIDs.GCL(silly = "PHOGAN19", IDs = PHOGAN19_hets_remove)

ProjectSillys_Origin_SampleSizes[, "Heterozygosity"] <-  ProjectSillys_Origin_SampleSizes[, "Genotyped"] - ProjectSillys_Origin_SampleSizes[, "Missing"] - ProjectSillys_Origin_SampleSizes[, "Duplicate"] - sapply(paste0(ProjectSillys_Origin, ".gcl"), function(x) get(x)$n)
```

## Final dataset
How many fish did we end up with?
```{r qa_final}
ProjectSillys_Origin_SampleSizes[, "Final"] <- sapply(paste0(ProjectSillys_Origin, ".gcl"), function(x) get(x)$n)
ProjectSillys_Origin_SampleSizes
save_objects("ProjectSillys_Origin_SampleSizes", path = "../../Objects/Hogan_17_19/")
```
Final dataset contains 8,289 total fish: 4,089 from 2017 and 4,197 from 2019.

Let's save the post-QA genotypes.
```{r save_postQA_genotypes}
if(!dir.exists("../../Genotypes/Hogan_17_19/OceanAK_Origin_PostQA")) {dir.create("../../Genotypes/Hogan_17_19/OceanAK_Origin_PostQA")}
save_sillys(ProjectSillys_Origin, "../../Genotypes/Hogan_17_19/OceanAK_Origin_PostQA/")
```

# Export genepop file
Now that we have final post QA/QC genotypes, we want to save our work by exporting the genotypes as a **Genepeop** file using `gcl2Genepop.GCL`, with individuals named by *SillySource*. That **Genepop** file will then feed in to Chase's *genepop2franz.GCL.r* script to create a **FRANz** input file for parentage analysis.
```{r genepop}
if(!dir.exists("../../Genepop/Hogan_17_19")) {dir.create("../../Genepop/Hogan_17_19")}

gcl2Genepop.GCL(sillyvec = c("PHOGAN17n", "PHOGAN17h", "PHOGAN19"), 
                loci = loci, 
                path = "../../Genepop/Hogan_17_19/Hogan_17_19_postQA.gen", 
                VialNums = TRUE, 
                usat = FALSE)
```

# Convert genepop to Franz
```{r Convert genepop to Franz for parentage}
detach(package:reshape, unload = TRUE)
source("../genepop2franz.GCL.r")
genepop2franz.GCL(
  Genepop = "../../Genepop/Hogan_17_19/Hogan_17_19_postQA.gen",
  OceanAK = "../../OceanAK/AHRP Salmon Biological Data 20210115_100431_names.csv",
  Year = c(2017, 2019),
  Stream = "HOGAN",
  output_dir = "../"
)
```
End